{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "67a545a6-86ef-4118-b6be-cc980a4b45a5",
   "metadata": {},
   "source": [
    "First we need to set up the environment and load the packages we will use for this workshop."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33da7aef-01f2-4d2b-9a15-3b1a2ba9a820",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.io import mmread\n",
    "from anndata import AnnData\n",
    "import squidpy as sq\n",
    "print(f\"squidpy=={sq.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58962e23-8657-401d-be31-3fb6c01952c0",
   "metadata": {},
   "source": [
    "Sets the path to the directory containing the Xenium output data - this is the directory where all of the outputs are stored."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4841334-9494-4b7a-b895-c8167ced94ee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_dir = '/nvme/project/shared/r/3_r_spatial/DATA/XENIUM_COLON_SUBSET'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76973199-e27f-468b-b20c-b1461ce2c0de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "os.listdir(data_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42aa5e7d-776b-4128-8897-9ff3332d3acb",
   "metadata": {},
   "source": [
    "Typically, you would use a loader for your specific data type to load data here - e.g. [spatialdata_io.xenium()](https://spatialdata.scverse.org/projects/io/en/latest/generated/spatialdata_io.xenium.html).\n",
    "\n",
    "However, often the data files from different platforms are still changes and the readers often do not work, or are not up to date.\n",
    "\n",
    "Therefore, we will cover how to read in the individual pieces of data and assemble a seurat object without the help of a loader.\n",
    "\n",
    "The main components that we need are:\n",
    "\n",
    "1. cell x gene matrix\n",
    "2. cell meta data\n",
    "3. cell coordinates - centroids and/or segmentation boundaries\n",
    "4. (optional) transcript coordinates\n",
    "\n",
    "We start by loading the cell x gene matrix, including feature and barcode metadata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56953f6e-83c6-44b7-984c-05fc4901311f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "features = pd.read_table(os.path.join(data_dir, \"cell_feature_matrix\", \"features.tsv.gz\"), header=None, names=[\"gene_ids\", \"gene_name\", \"feature_types\"]).set_index(\"gene_name\")\n",
    "features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab2e4822-8072-40d9-81b4-091ecef457a8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "barcodes = pd.read_table(os.path.join(data_dir, \"cell_feature_matrix\", \"barcodes.tsv.gz\"), header=None,  names=[\"cell_id\"])\n",
    "barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dacbe929-32b7-4cae-ad99-6b803c1b4270",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "matrix = mmread(os.path.join(data_dir, \"cell_feature_matrix\", \"matrix.mtx.gz\"), spmatrix=True).tocsr().T\n",
    "matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd02078d-10e1-4e1b-baf4-57130299c848",
   "metadata": {},
   "source": [
    "Read in additional information about the cells - this gives us pre-calculated information, for example segmented cell or nucleus size for each cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77ced96f-d9bc-470c-9b82-5a0a9f036cab",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cells = pd.read_csv(os.path.join(data_dir, \"cells.csv.gz\"))\n",
    "cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25b74393-6273-42d8-8b82-8260d7d9bda2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cells[['x_centroid', 'y_centroid']].agg(['min', 'max'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35a61e00-9946-40bd-8067-d42f0074139b",
   "metadata": {},
   "source": [
    "Next, we read in cell boundaries:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa8906bf-16fe-40be-9be4-ecf78346c74b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cell_polygons = pd.read_csv(os.path.join(data_dir, \"cell_boundaries.csv.gz\"))\n",
    "cell_polygons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5422085b-b387-4857-8570-0cc40d6de118",
   "metadata": {},
   "source": [
    "While one would expect that "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6af276f8-fa5f-41fd-83a3-c31efaeaf57e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# rotate the coordinates to match the orientation of Seurat plots\n",
    "# however, do not flip the Y-axis because we'll use a function that already orientates the Y-axis correctly\n",
    "# remember this when you come across gdf.plot() later\n",
    "cell_polygons.columns = ['cell_id', 'vertex_y', 'vertex_x']\n",
    "cell_polygons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79f641e1-d888-42ce-91cb-7f014d2bb364",
   "metadata": {},
   "source": [
    "And individual transcript coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61d8a15a-3e5f-411d-9e41-906c69acf949",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "transcripts = pd.read_csv(os.path.join(data_dir, \"transcripts.csv.gz\"))\n",
    "transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c03cac0e-fbfd-4b45-ae03-88fff51e3ca8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# rotate the coordinates to match the orientation of Seurat plots\n",
    "transcripts_rotated = pd.DataFrame({\n",
    "    'x_location': transcripts[\"y_location\"],\n",
    "    'y_location': transcripts[\"x_location\"]\n",
    "})\n",
    "transcripts[['x_location', 'y_location']] = transcripts_rotated\n",
    "transcripts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7d68025-2124-4621-8cb8-830bcc4feebe",
   "metadata": {},
   "source": [
    "Then, we will assemble gene expression matrices. Matrix file is further split into gene expression matrix and various control probes and codewords. Different platforms and platform versions include different control probes. As this will vary, it’s important to check and understand what the specific controls in your own data are.\n",
    "\n",
    "Here, negative control probes are probes that are added to the reaction but target non-biological sequences and should not bind any tissue RNA. Negative control codewords are valid codewords, but no probes with that codeword added to the reaction. This effectively tells us how good the transcript calling algorithm is.\n",
    "\n",
    "For convenience, we will split control probes and codewords into separate assays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3849dd91-1437-4afc-a692-f2a644688474",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gex_matrix = matrix[:, (features['feature_types'] == \"Gene Expression\").values]\n",
    "gex_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edbe04e4-3312-4f88-9f73-de8a45124264",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ncp_matrix = matrix[:, (features['feature_types'] == \"Negative Control Probe\").values]\n",
    "ncp_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f3838ea-3ca1-4696-af11-dcaf331a5100",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ncc_matrix = matrix[:, (features['feature_types'] == \"Negative Control Codeword\").values]\n",
    "ncc_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ffce689-4608-4dfc-b664-6cd69cdf7af4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "uc_matrix = matrix[:, (features['feature_types'] == \"Unassigned Codeword\").values]\n",
    "uc_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "353bc2f4-0dc8-4a49-af38-c5a47dfcb06c",
   "metadata": {},
   "source": [
    "Then, we assemble a basic `AnnData` object with gene expression matrix and metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a3c230d-bb9c-465d-adeb-7ac7d76a8d8f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata = AnnData(X=gex_matrix, obs=cells)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5866d73b-2a07-457f-adf5-c4f422eae7ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs_names = adata.obs[\"cell_id\"]\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fac315c-897b-4978-8445-77084b9d21f9",
   "metadata": {},
   "source": [
    "Add gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42643da9-191c-411b-8ac4-cad6ce91395d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.var_names = features[features[\"feature_types\"] == \"Gene Expression\"].index\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "975a9471-50f5-447b-bdf0-d9819c7fda1d",
   "metadata": {},
   "source": [
    "We add the spatial coordinates of cells as documented [here](https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_read_spatial.html#spatial-coordinates-in-anndata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc1d9d09-1c88-4415-93db-0f483d764253",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# rotate the coordinates to match the orientation of Seurat plots\n",
    "centroids = cells[[\"x_centroid\", \"y_centroid\"]].to_numpy()[:, [1, 0]]\n",
    "centroids[:, 1] = -centroids[:, 1]\n",
    "centroids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eea2451-756f-431f-894c-c6c94a759192",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obsm[\"spatial\"] = centroids\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb43a508-b6eb-4826-84dc-4e24a9fecc4f",
   "metadata": {
    "tags": []
   },
   "source": [
    "Let’s start with some basic QC and visualisation of the data.\n",
    "\n",
    "First, lets visualise the total transcripts detected per cell.\n",
    "\n",
    "As in scRNA-Seq data, this is the most basic measure of overall signal and how well the data looks.\n",
    "\n",
    "Unlike in scRNA-Seq data or unbiased sequencing-based ST, these measures are also very heavily dependent not only on the total RNA quantity of each cell and tissue quality, but also on the target panel used for the experiment. Under-represented cell types will naturally yield fewer transcripts. Finally, the quality of cell segmentation also plays a role.\n",
    "\n",
    "In this case, we can see that there are areas with higher and lower total transcripts detected.\n",
    "\n",
    "Understanding your tissue and target panel here is important to delineate where these differences are biological and where they may be technical."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387574d6-42e9-4083-8997-5018a08563c9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=[\n",
    "        \"transcript_counts\",\n",
    "    ],\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f91198f3-53fa-4ba9-b5a1-187f14c4df98",
   "metadata": {},
   "source": [
    "Similarly, we can visualise the total number of gene detected per cell. You can see that this is a bit less variable across tissue.\n",
    "\n",
    "This can also suggest that there cells at the top of the epithelial crypts in this sample with genes detected at high copy number than the rest of the tissue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2391a3c2-7ff1-4bb8-b319-b639eed6da81",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"n_features\"] = (adata.X > 0).sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5cd3857-cf4a-41a1-bcc2-7902f22964ce",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=[\n",
    "        \"n_features\",\n",
    "    ],\n",
    "    wspace=0.4,\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47e586c4-35f8-4821-8728-d32ccad40679",
   "metadata": {
    "tags": []
   },
   "source": [
    "This code examines the distribution of the number of features (genes) detected per cell in the AnnData object\n",
    "using a density plot and calculates specific quantiles of this distribution.\n",
    "This is important for understanding the variability and distribution of detected features,\n",
    "which can help identify potential issues such as low-quality cells and determine any filtering thresholds\n",
    "that may need to be applied.\n",
    "\n",
    "If you’re coming from scRNA-Seq work, these low numbers probably look very alarming.\n",
    "How can you possibly work with 31 median genes per cell?\n",
    "\n",
    "Unlike scRNA-Seq data and sequencing-based ST, both gene dropouts and noise are much, much lower in in situ ST data.\n",
    "\n",
    "We are also working with 100-fold fewer targetted genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8410447-d5b6-4369-b71e-4563f0a6e35c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.kdeplot(data=adata.obs, x=\"n_features\", fill=True)\n",
    "plt.xlabel(\"n_features\")\n",
    "plt.ylabel(\"Density\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dc11a4c-4975-4677-b5eb-d90b39219149",
   "metadata": {
    "tags": []
   },
   "source": [
    "We can also report quantiles of that same variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7959dcfe-8ade-4ca0-98d8-050bbc4ee33c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"n_features\"].quantile([0.01, 0.10, 0.50, 0.90, 0.99])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dffbdb5-23aa-44eb-97e3-1071a6c44b2e",
   "metadata": {},
   "source": [
    "Using `sq.pl.spatial_scatter()ImageFeaturePlot(seurat, \"cell_area\") + scale_fill_viridis_c()\n",
    "` to visualize the cell area in spatial transcriptomics data allows us to examine the spatial organization and potential heterogeneity of cell sizes within your tissue sample.\n",
    "\n",
    "Why do we get such a difference in spatial distribution of cell sizes?\n",
    "\n",
    "This could be due to biological differences between small and large cells - e.g. small cells like T-cells.\n",
    "\n",
    "However, here the signal correlates with areas of low cellularisation. Therefore, it is likely this is an artefact of nuclei expansion in cell segmentation.\n",
    "\n",
    "What is Nuclei Expansion?\n",
    "\n",
    "Nuclei expansion in cell segmentation refers to the process of enlarging the segmented nuclei regions to approximate the boundaries of the entire cells. This technique is used to better represent the actual cell boundaries when only the nuclei have been explicitly segmented/we only have DAPI and no additional cell boundary staining. The primary goal is to provide a more accurate estimation of the cellular area, which is crucial for various downstream analyses in spatial transcriptomics and single-cell studies. In this case, nuclei expansion is constrained either by maximum distance or other nearby cells - so, where there are no other nearby cells to “bump into”, the expansion generates artificially bigger cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8baa7ef7-39ab-4eff-9526-ed67dcab2e54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ImageFeaturePlot(seurat, \"cell_area\") + scale_fill_viridis_c()\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=[\n",
    "        \"cell_area\",\n",
    "    ],\n",
    "    wspace=0.4,\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b89dd71b-164f-4ed6-86dd-f792728273ed",
   "metadata": {},
   "source": [
    "We can further check that this is likely the case by plotting the ratio between nuclei and total cell area. We can see that there is a very big decrease in percentage of cell area occupied by nucleus in areas of low cell density.\n",
    "\n",
    "The cell-to-nucleus area ratio can also potentially provide insights into cell morphology, cell type and potential changes in cellular states or conditions. For example, T-Cells can often be quite well identified by this variable alone, as they have a small cytoplasm volume. However, without a cell boundary stain, this metric mainly captures segmentation artefacts, so be careful about over-interpretation!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43fd6789-9e89-4071-8dfe-3e4f2a25b577",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"cell_nucleus_ratio\"] = adata.obs[\"nucleus_area\"] / adata.obs[\"cell_area\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5929df0f-cb0e-4e35-86c2-789f148430e2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ImageFeaturePlot(seurat, \"cell_area\") + scale_fill_viridis_c()\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=[\n",
    "        \"cell_nucleus_ratio\",\n",
    "    ],\n",
    "    wspace=0.4,\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8c2f773-f351-4962-a425-0cb091d53345",
   "metadata": {},
   "source": [
    "If we look at the distribution, we see that we have a big tail end of overly large cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25023e8a-520c-4825-a6c0-a22d49baec23",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.kdeplot(data=adata.obs, x=\"cell_area\", fill=True)\n",
    "plt.xlabel(\"cell_area\")\n",
    "plt.ylabel(\"Density\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08f1ac49-a107-4afb-b33c-413409712793",
   "metadata": {},
   "source": [
    "In this case, we can see that as expected, there is generally a correlation between cell area and transcript detection rate.\n",
    "\n",
    "However, we also have a group of cells where this is not the case - very large cells but relatively few transcripts. These cells are mainly submucosal stromal cells which are very poorly covered by the panel 10x have used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2717099-fee5-44e2-afc6-4a3d355c59cc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.scatterplot(data=adata.obs, x=\"total_counts\", y=\"cell_area\", s=5)\n",
    "plt.xlabel(\"total_counts\")\n",
    "plt.ylabel(\"cocell_areal2\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad59d389-e36d-43db-ad31-e0e4f23d53ab",
   "metadata": {
    "tags": []
   },
   "source": [
    "We can create a filter to remove the overly large cells from the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c16eef3-12cd-4fea-9a5b-b607fd0e6915",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"size_filter_large\"] = adata.obs[\"cell_area\"] < np.quantile(adata.obs[\"cell_area\"], 0.99)\n",
    "# re-type as categorical (useful for plotting) \n",
    "adata.obs[\"size_filter_large\"] = adata.obs[\"size_filter_large\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1622f8e-088a-4351-9f69-e1375c8c9418",
   "metadata": {},
   "source": [
    "Now we can use `sq.pl.spatial_scatter()` to visualise the cells which have been flagged for removal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "778f00bb-65a0-41f5-b6d7-9070565733ca",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"size_filter_large\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b76bf9b-7d48-493d-b97c-09de7b34b023",
   "metadata": {},
   "source": [
    "We can use the same approach to create a filter for segmented cells which are very small and likely segmentation arfetacts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b3398e2-b87e-4d17-a9f9-174e1bade2e1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"size_filter_small\"] = adata.obs[\"cell_area\"] > np.quantile(adata.obs[\"cell_area\"], 0.01)\n",
    "# re-type as categorical (useful for plotting) \n",
    "adata.obs[\"size_filter_small\"] = adata.obs[\"size_filter_small\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cc06513-90d1-4f5c-8e60-3bb278f62793",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"size_filter_small\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9157866-1833-4346-b983-d32b93fa7679",
   "metadata": {},
   "source": [
    "We can check how these values correlate with gene detection rate.\n",
    "\n",
    "If we filter out small cells, we will remove cells with low numbers of genes detected.\n",
    "\n",
    "If we filter out large cells, this is not that biased towards overly large counts, as we saw before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ca327a3-e712-4588-8ebd-9f49a94c96f5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.violinplot(data=adata.obs, x=\"size_filter_small\", y=\"n_features\", palette=\"tab10\")\n",
    "# sns.swarmplot(data=adata.obs, x=\"size_filter_small\", y=\"n_features\", color=\"black\", size=3)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cbbd0bc-9bb2-4dbf-afd6-1e2b560715cf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.violinplot(data=adata.obs, x=\"size_filter_large\", y=\"n_features\", palette=\"tab10\")\n",
    "# sns.swarmplot(data=adata.obs, x=\"size_filter_large\", y=\"n_features\", color=\"black\", size=3)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0461bfc-9e5d-4dc3-b90f-1da57cbdef3f",
   "metadata": {},
   "source": [
    "Adjusting the threshold for what is considered a “small cell” can have significant implications for your analysis, especially in areas with specific cell types such as T-cells, which are small and densely packed in follicular regions. This example demonstrates how changing the threshold to the 10th percentile affects the filtering. In this case, we would probably filter out a lot of good cells that we don’t want to lose! So, be careful when looking at these types of QC metrics!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e853bc58-3c40-4cb9-ab7c-a873a0689f60",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"size_filter_small\"] = adata.obs[\"cell_area\"] > np.quantile(adata.obs[\"cell_area\"], 0.1)\n",
    "# re-type as categorical (useful for plotting) \n",
    "adata.obs[\"size_filter_small\"] = adata.obs[\"size_filter_small\"].astype(\"category\")\n",
    "# plot\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"size_filter_small\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67c33eb3-daff-4aec-b075-b614dc765f0b",
   "metadata": {},
   "source": [
    "Lets set this back to the original 1% threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4824f5b-5808-454b-b0ac-986df92be38b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"size_filter_small\"] = adata.obs[\"cell_area\"] > np.quantile(adata.obs[\"cell_area\"], 0.01)\n",
    "# re-type as categorical (useful for plotting) \n",
    "adata.obs[\"size_filter_small\"] = adata.obs[\"size_filter_small\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e611ad-d451-444e-8d03-179c803535b6",
   "metadata": {},
   "source": [
    "The most important filter is the overall transcript detection. Empty cells or cells with very low transcript count cannot be taken forward for clustering analysis and it is extremely difficult to identify what they may be. Here, we set a threshold of minimum 15 transcripts. This seems quite low - for data from in situ platforms with low noise (Xenium, Merfish, Merscope), this is generally enough to cluster and identify cell types. If your data has more noise (e.g. CosMx), a higher threshold is more appropriate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa646b67-37bb-4ee6-95eb-d6c55da187d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"transcript_filter\"] = adata.obs[\"total_counts\"] >= 15\n",
    "# re-type as categorical (useful for plotting) \n",
    "adata.obs[\"transcript_filter\"] = adata.obs[\"transcript_filter\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11086c7c-7d9e-42ff-b939-2ec03608821c",
   "metadata": {},
   "source": [
    "And we can visualise the cells that we would lose.\n",
    "\n",
    "We see that we disproportionately would filter out more cells from some regions than others. As pointed out previously, this is likely due to a combination of gene panel coverage in some regions and very small cells in densely packed regions like follicles.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eec4888-c40c-4f3c-8d07-8b4437552050",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"transcript_filter\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d464cab7-5789-4798-ba7f-5f5c72b5314d",
   "metadata": {},
   "source": [
    "Finally, visualizing the counts of negative control codewords, negative control probes, and unassigned codewords helps identify and understand technical artifacts and background noise in your spatial transcriptomics data.\n",
    "\n",
    "Here, we can see that all control probes and codewords produce yield very little signal, suggesting our data is good quality!\n",
    "\n",
    "In some cases, high amount of autoflourescence is the cells/tissue can sometimes generate false positive signal and this should be filtered out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "529e67d5-d5d8-4554-abae-005c48c0afd9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"n_features_neg_ctrl_codeword\"] = ncc_matrix.sum(axis=1)\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"n_features_neg_ctrl_codeword\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ea26787-0c18-4a6a-84e6-244c0ca242d5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"n_features_neg_ctrl_probe\"] = ncp_matrix.sum(axis=1)\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"n_features_neg_ctrl_probe\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7d59f4d-c1ad-492e-875f-9845338096a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"n_features_unassigned_codeword\"] = uc_matrix.sum(axis=1)\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"n_features_unassigned_codeword\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd22d135-5160-4b12-93e8-accab1c15412",
   "metadata": {},
   "source": [
    "Although the negative control signal is low, we can nonetheless create a filter to remove cells which have any, although in this case it is probably unnecessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f4e6490-c3b0-4dd9-a091-12ffe0b788c0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"probe_filter\"] = (\n",
    "    (adata.obs[\"n_features_unassigned_codeword\"] == 0) &\n",
    "    (adata.obs[\"n_features_neg_ctrl_codeword\"] == 0) &\n",
    "    (adata.obs[\"n_features_neg_ctrl_probe\"] == 0)\n",
    ")\n",
    "# re-type as categorical (useful for plotting) \n",
    "adata.obs[\"probe_filter\"] = adata.obs[\"probe_filter\"].astype(\"category\")\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"probe_filter\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a532bd17-7080-415a-9881-4c83aa3df575",
   "metadata": {},
   "source": [
    "Finally, we can subset the `AnnData` object based on any/all of the filters we have created earlier.\n",
    "\n",
    "By combining probe, size, and transcript filters, you can retain only the cells that meet all quality criteria, reducing the impact of technical artifacts and noise on your analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26d93665-6459-49d3-8a73-794bd9331482",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# for readability, combine all the filters into one\n",
    "mask = (\n",
    "    (adata.obs[\"probe_filter\"].astype(bool)) &\n",
    "    (adata.obs[\"size_filter_large\"].astype(bool)) &\n",
    "    (adata.obs[\"size_filter_small\"].astype(bool)) &\n",
    "    (adata.obs[\"transcript_filter\"].astype(bool))\n",
    ")\n",
    "adata = adata[mask].copy()\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10fc9dc2-c8b4-4940-ad22-356963e865ce",
   "metadata": {},
   "source": [
    "## Data Normalisation\n",
    "\n",
    "Normalize counts per cell using `scanpy.pp.normalize_total`.\n",
    "\n",
    "Per [squidpy xenium tutorial](https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_xenium.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3551b625-1e53-47ff-b293-8028cf8f9aa6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "sc.pp.normalize_total(adata, inplace=True)\n",
    "sc.pp.log1p(adata)\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d34c1922-cd41-4e2c-a601-dfa76b6f4bb1",
   "metadata": {},
   "source": [
    "Principal Component Analysis (PCA) is a dimensionality reduction technique used to identify the primary axes of variation in high-dimensional data. In the context of spatial transcriptomics, PCA helps to reduce the complexity of the data while preserving the most important patterns of variation.\n",
    "\n",
    "TIP: If your target panel is very small, you can skip this step and carry out clustering analysis directly on gene expression. This can sometimes help with achieving better clustering results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fb0c444-2f48-4d95-8c11-15b595f41e63",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.pca(adata)\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdaa95a-22e1-4d72-b803-ac9a94184f93",
   "metadata": {
    "tags": []
   },
   "source": [
    "As before, we can visualise how much variation is captured by each PC.\n",
    "\n",
    "The information is stored in `adata.uns['pca']['variance_ratio']`.\n",
    "\n",
    "An elbow plot helps to determine the number of significant PCs to use for downstream analyses.\n",
    "The plot typically shows the amount of variance explained by each PC, and the “elbow” point indicates a natural cutoff."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c7d2fa8-6a82-44b7-a4d7-b6fc54fbccbf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "vr = adata.uns['pca']['variance_ratio']\n",
    "\n",
    "plt.figure(figsize=(8, 5))\n",
    "plt.plot(range(1, len(vr)+1), vr, marker='o')\n",
    "plt.xlabel(\"Principal Component\")\n",
    "plt.ylabel(\"Variance Explained Ratio\")\n",
    "plt.title(\"Scree Plot\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4146350-da27-4404-8438-1411ed986213",
   "metadata": {},
   "source": [
    "Plotting the top genes contributing to a specific principal component helps in understanding the biological factors driving the variation captured by that component. This type of plot highlights the genes with the highest loadings, which are the most influential in the principal component analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bff328f-858b-43d0-8aac-e91c8ca4fadc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.pca_loadings(adata, components='1,2')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3074d2c-5ccb-4795-9c30-7fdf68a7f681",
   "metadata": {},
   "source": [
    "The `sc.pl.pca()` function in Scanpy is used to visualize the expression of a specific gene across cells in a given dimensionality reduction space (e.g., PCA). This helps to understand how the expression of a gene varies across the principal components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cb39f45-c0d9-4c13-84cf-6a6e44a83665",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.pca(adata, color=\"CEACAM5\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1989f7ab-34ec-45c4-a8a4-72205c484aa5",
   "metadata": {},
   "source": [
    "We can also examine how various PCs are distributed spatially.\n",
    "\n",
    "Here, we can see that high PC1 loadings enrich in crypt top cells and low PC1 loadings enrich in follicular structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31e7f4f2-98cf-4d29-b33c-0b561a710a7b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"PC1\"] = adata.obsm[\"X_pca\"][:, 0]\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"PC1\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc290652-961e-4d51-b0a4-b73af6a6644a",
   "metadata": {
    "tags": []
   },
   "source": [
    "We can plot the expression of high (or low) loading genes to visualise how this correlates with our dimensionality reduction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dde4b90-bca7-4bb8-892e-c2599c28faf4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"MS4A1\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a40824d4-2dc6-43bb-bd55-4491af663df6",
   "metadata": {},
   "source": [
    "Next, we will use the reduced dimensionality data for clustering and cluster visualisation.\n",
    "\n",
    "`sc.pp.neighbors()` needs to be run prior to `sc.tl.umap()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f5a9117-644e-41b2-ac0a-a0dcfc95c597",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata)\n",
    "sc.tl.umap(adata)\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eac2fe8-8f79-4111-94c2-d82a5e66a243",
   "metadata": {},
   "source": [
    "Then we can also run `sc.tl.leiden()` to cluster cells using the Leiden algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1751049-f220-40f5-9ebe-ffdc48d57cf8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata)\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98282e87-fa86-4244-8ec4-69d817c3ea66",
   "metadata": {},
   "source": [
    "Next lets visualise the clusters - firstly, based on transcriptome embedding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdd1675d-02b3-46ac-8280-f4b475b77c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"leiden\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdf2b6ab-efb8-419a-bc04-0758a4e009d7",
   "metadata": {},
   "source": [
    "And now lets plot the clusters in tissue space.\n",
    "\n",
    "We can see that our clusters have quite nice correspondence to distinct spatial regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fd418b8-c1f1-42a6-89d3-05008efc7b09",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"leiden\",\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e26c16b1-fc19-4bf8-b9ad-f4b9614c0158",
   "metadata": {},
   "source": [
    "As before, now we can use scanpy differential expression functions to identify marker genes for specific cell clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28f7e073-9703-45b8-b750-53fd9dcf1d9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, 'leiden', method='t-test', groups = ['0'])\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce941bff-464e-431c-a7b6-97c74b2f2a36",
   "metadata": {},
   "source": [
    "We can visualise expression of cluster specific markers using feature plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5306b9d7-b59b-4ab4-8215-e51781776c9f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"CD3E\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93652d24-c5bf-42cb-a7b0-92b7f8e2c673",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"MS4A1\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90c86f7b-9650-4181-b094-c6aefe85098c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"CEACAM5\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9275aa02-f198-45b3-86f6-59017b835208",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"KIT\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e23f1ad-8180-4a9c-814e-a13b5489f676",
   "metadata": {},
   "source": [
    "Or, as in our sequencing ST tutorial, detect and visualise top markers for every cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b122acb9-b502-4253-882d-fcfeb773be93",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, 'leiden', method='t-test')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1531d6f-81e4-4d5a-9b07-e5d9d0ed67f8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "top = []\n",
    "# adata.obs[\"leiden\"].cat.categories.size\n",
    "for i in adata.obs[\"leiden\"].cat.categories.tolist():\n",
    "    top.extend(sc.get.rank_genes_groups_df(adata, i).head(5)[\"names\"].to_list())\n",
    "top = list(dict.fromkeys(top))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47771e46-3e96-407d-b0f0-6c810596311c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "top = pd.DataFrame(columns=[\"gene\", \"cluster\"])\n",
    "for i in adata.obs[\"leiden\"].cat.categories.tolist():\n",
    "    i_top = sc.get.rank_genes_groups_df(adata, i).head(5)[\"names\"].to_list()\n",
    "    i_new = pd.DataFrame({\n",
    "        \"cluster\": [i] * len(i_top),\n",
    "        \"gene\": i_top\n",
    "    })\n",
    "    top = pd.concat([top, i_new], ignore_index=True)\n",
    "top.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "064b3a84-3e1b-43e1-8e9b-49f6048b16cb",
   "metadata": {},
   "source": [
    "`sc.pl.dotplot()` provides a convenient and visually appealing way to display expression patterns of top marker genes across clusters using a dot plot.\n",
    "\n",
    "Note for : it would be nice to rotate this plot 90 degrees,\n",
    "to match [Clustered_DotPlot()](https://samuel-marsh.github.io/scCustomize/reference/Clustered_DotPlot.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2f7f1ed-0e3f-47ad-b5f8-df085e3b88d7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "starts = np.arange(0, top.shape[0], 5)\n",
    "ranges = [(s, s + 4) for s in starts]\n",
    "sc.pl.dotplot(\n",
    "    adata, top[\"gene\"], \"leiden\",\n",
    "    swap_axes=True,\n",
    "    var_group_positions=ranges,\n",
    "    var_group_labels=adata.obs[\"leiden\"].cat.categories.tolist()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "234bef72-de1b-4727-9b44-fc14736120a2",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Additional Spatial Visualisations\n",
    "\n",
    "The resolution of in situ datasets is typically very high and so it can be difficult to visualise everything in one plot. Below, we will explore different visualisations that can help unpick and understand the data a bit better.\n",
    "\n",
    "To better visualise spatial distribution of clusters, sometimes it can be useful to subset only certain groups to reduce crowding. Here, we specifically only visualising two selected clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8b8fdd3-0c72-43e6-b243-e923750d59c0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"leiden\",\n",
    "    groups=[\"0\", \"1\"],\n",
    "    figsize = (12, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe93c73f-a11c-461a-ac40-e4701dd40f38",
   "metadata": {},
   "source": [
    "Sometimes, it can be useful to create additional fields of view of the data - for example, zooms of specific regions. First, let’s look at the coordinate system by plotting the data and turning on the plotting of the axes, which are off by default to create nicer looking plots.\n",
    "\n",
    "This gives us a rough idea on where in the coordinate system to create any subsets or zooms of the data.\n",
    "\n",
    "For example, if we want to zoom in on the follicle in the top right corner, we can see that it lies roughly between 4000:5000 and -8000:-9000 coordinate regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec120001-b731-4fb0-9e3b-e6ad81c937df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "x_min = round(adata.obsm[\"spatial\"][:, 0].min() / 1000)\n",
    "x_max = round(adata.obsm[\"spatial\"][:, 0].max() / 1000) + 1\n",
    "y_min = round(adata.obsm[\"spatial\"][:, 1].min() / 1000)\n",
    "y_max = round(adata.obsm[\"spatial\"][:, 1].max() / 1000) + 1\n",
    "ax = sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"leiden\",\n",
    "    return_ax=True,\n",
    ")\n",
    "ax.set_xticks(np.arange(x_min * 1000, x_max * 1000, 1000))\n",
    "ax.set_yticks(np.arange(y_min * 1000, y_max * 1000, 1000))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7e7971-dc91-42db-8791-3e6309e6b910",
   "metadata": {},
   "source": [
    "We can then use the `crop_coord` argument, specifying the `(xmin, ymin, xmax, ymax)` coordinates of the region of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4df8bf8-3ed6-42b2-a403-bdbb03bdba25",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"leiden\",\n",
    "    crop_coord=(4000, -9000, 5000, -8000)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "951e8948-c32e-4a45-b314-b32f22e24f25",
   "metadata": {},
   "source": [
    "As we are zooming in closer to the tissue, we can also switch from plotting cell centroids (i.e. dots) by default to visualising cell segmentation boundaries.\n",
    "Plotting cell boundary polygons for large FOVs can be quite time consuming, and doesn’t provide much more detail on a fully zoomed-out view.\n",
    "\n",
    "For this, we need the coordinates of the cell boundaries and a fair amount of custom code to format and plot the cell boundaries."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45559260-7303-471b-8d38-f7d2d8bfb331",
   "metadata": {},
   "source": [
    "First, we'll make a copy of the `adata` object for the region of interest, which we'll use to identify the cells to display in the segmentation plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c339eb4a-695e-4f0a-9592-43d417266762",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "spatial = adata.obsm['spatial']\n",
    "\n",
    "xmin, xmax = 4200, 5000\n",
    "ymin, ymax = -8800, -8000\n",
    "\n",
    "mask = (\n",
    "    (spatial[:,0] > xmin) & (spatial[:,0] < xmax) &\n",
    "    (spatial[:,1] > ymin) & (spatial[:,1] < ymax)\n",
    ")\n",
    "\n",
    "adata_cropped = adata[mask].copy()\n",
    "adata_cropped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ed1167c-b91a-4d17-b2a1-53ecbc16c57b",
   "metadata": {},
   "source": [
    "Next, we'll a dictionary where each key is a cell identifier and the corresponding value is a `Polygon` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4702c18a-ca6d-4567-814b-ebe34cf13dc8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from shapely import Polygon\n",
    "\n",
    "def create_cell_polygons(\n",
    "    cell_polygons_df: pd.DataFrame,\n",
    "):\n",
    "    cell_shapes = {}\n",
    "    \n",
    "    for cell_id, group in cell_polygons_df.groupby('cell_id'):\n",
    "        cell_points = group[['vertex_x','vertex_y']].values\n",
    "        if len(cell_points) > 2:\n",
    "            cell_shapes[cell_id] = Polygon(cell_points)\n",
    "    return cell_shapes\n",
    "   \n",
    "cell_shapes = create_cell_polygons(\n",
    "    cell_polygons\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e995934-d905-4cd9-8af2-94030cf93fef",
   "metadata": {},
   "source": [
    "For instance, here is how a given cell can be accessed, and how it looks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c8f5717-f609-423b-9225-2661e8e2eba6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cell_shapes['adcbkhka-1']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd2fde2-2745-4bbc-bd78-1200e3e78998",
   "metadata": {},
   "source": [
    "We can then create a `GeoDataFrame` that contains only the subset of cells within the selected field of view.\n",
    "\n",
    "Note: the filtering of cells is based on the cell centroid. Some cells that partially overlap the field of view will not be visible due to that filter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e68d24fe-62f7-4087-b50e-4a75f33322c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import geopandas as gpd\n",
    "\n",
    "gdf = gpd.GeoDataFrame(\n",
    "    {\"cell_id\": list(cell_shapes.keys()),\n",
    "     \"geometry\": list(cell_shapes.values())}\n",
    ").set_index(\"cell_id\")\n",
    "\n",
    "common_ids = gdf.index.intersection(adata_cropped.obs_names)\n",
    "gdf = gdf.loc[common_ids]\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "848c0115-a40f-46bc-ade3-28beba216a1a",
   "metadata": {},
   "source": [
    "We can then plot the segmented field of view with `gdf.plot()` and some more custom code to colour by cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15f66a44-e0b8-4bc3-a632-ec77b9e290a3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8,8))\n",
    "\n",
    "gdf.plot(ax=ax, edgecolor='black', linewidth=1)\n",
    "gdf2 = gdf.join(adata_cropped.obs[['leiden']])\n",
    "gdf2.plot(ax=ax, column='leiden', cmap='hsv', alpha=0.6, legend=True)\n",
    "\n",
    "ax.set_facecolor(\"black\")\n",
    "ax.set_aspect('equal')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f273ec3b-7b30-4831-b803-b6d04add7806",
   "metadata": {},
   "source": [
    "We can visualise gene expression or other continous variable on the new FOV as before.\n",
    "\n",
    "For example, here we have MS4A1/CD20 expression, which is a B-Cell marker. We can see it quite nicely limited to the lymphoid follicle."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "507e6060-99c4-4187-80a2-6b09533f3e45",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "feature = \"MS4A1\"\n",
    "\n",
    "feature_series = pd.Series(\n",
    "    data=np.asarray(adata_cropped[:, feature].X.todense()).squeeze(),\n",
    "    index=adata_cropped.obs_names.astype(str),\n",
    "    name=feature\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,8))\n",
    "\n",
    "gdf.plot(ax=ax, edgecolor='black', linewidth=1)\n",
    "gdf2 = gdf.join(feature_series, how=\"left\")\n",
    "gdf2.plot(ax=ax, column=feature, cmap='viridis', alpha=0.6, legend=True)\n",
    "\n",
    "ax.set_title(feature)\n",
    "ax.set_facecolor(\"black\")\n",
    "ax.set_aspect('equal')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea44ad66-a350-40c0-a147-36f668782aca",
   "metadata": {},
   "source": [
    "We can also overlay the coordinates of individual molecules to the plot. For example, here we are added some more T-cell and B-cell specific markers.\n",
    "\n",
    "This visualisation can be useful because molecules are stored independently of cells and cell boundaries.\n",
    "Therefore, if there are regions where cell segmentation is not good, or if cells were filtered out from clustering analysis due to their low quality, the molecules will remain and can still be visualised this way.\n",
    "\n",
    "For example, here we can see there are a few molecules of CXCR5 detected outside of cellular boundaries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a941921-c11f-459d-8644-3d2daf3b07ba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from matplotlib import cm\n",
    "\n",
    "feature = \"MS4A1\"\n",
    "molecules = [\"CXCR5\", \"FOXP3\"]\n",
    "\n",
    "feature_series = pd.Series(\n",
    "    data=np.asarray(adata_cropped[:, feature].X.todense()).squeeze(),\n",
    "    index=adata_cropped.obs_names.astype(str),\n",
    "    name=feature\n",
    ")\n",
    "\n",
    "transcripts_df = transcripts[transcripts['feature_name'].isin(molecules)].copy()\n",
    "# NOTE: we take the opposite of ymin and ymax due to the inverted y-axis\n",
    "mask = (\n",
    "    (transcripts_df[\"x_location\"] > xmin) & (transcripts_df[\"x_location\"] < xmax) &\n",
    "    (transcripts_df[\"y_location\"] > -ymax) & (transcripts_df[\"y_location\"] < -ymin)\n",
    ")\n",
    "transcripts_df = transcripts_df[mask]\n",
    "\n",
    "cmap_points = plt.colormaps.get_cmap(\"Set1\")\n",
    "gene_to_color = {g: cmap_points(i % 8) for i, g in enumerate(molecules)}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,8))\n",
    "\n",
    "gdf.plot(ax=ax, edgecolor='black', linewidth=1)\n",
    "gdf2 = gdf.join(feature_series, how=\"left\")\n",
    "gdf2.plot(ax=ax, column=feature, cmap='viridis', alpha=0.6, legend=True)\n",
    "\n",
    "for g in molecules:\n",
    "    sub = transcripts_df[transcripts_df['feature_name'] == g]\n",
    "    ax.scatter(sub['x_location'].values, sub['y_location'].values,\n",
    "               s=15, alpha=0.8,\n",
    "               c=[gene_to_color[g]], label=g, linewidths=0, zorder=3)\n",
    "\n",
    "ax.legend(title=\"molecules\", fontsize=\"small\")\n",
    "ax.set_title(feature)\n",
    "ax.set_facecolor(\"black\")\n",
    "ax.set_aspect('equal')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "283727f7-fd1d-4a38-a754-3b7838e4ffdf",
   "metadata": {},
   "source": [
    "## Additional exercises\n",
    "\n",
    "It has been demonstrated (Atta et al, 2024), that if you have a very skewed panel design in in situ data, cell area/size normalisation can be more robust.\n",
    "A cell with low panel coverage would have lower counts for example, so normalising to “library size” would inflate those counts.\n",
    "Note that for large panels (e.g. Xenim 5K), it’s not such an issue.\n",
    "\n",
    "Let’s try normalising by area instead - there’s no in-built method is Scanpy for this, so we add the normalised counts and log transformed values (for visualisation) back to the object manually:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8b946ba-e8b2-42ed-980d-cedb30f6dbcd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from scipy import sparse\n",
    "\n",
    "adata.layers['COUNT_NORM'] = adata.X\n",
    "raw_counts = adata.layers['counts']\n",
    "cell_area = adata.obs['cell_area'].values.astype(float) + 1e-6\n",
    "\n",
    "median_area = np.median(cell_area)\n",
    "\n",
    "if sparse.issparse(raw_counts):\n",
    "    # trick using dot product for sparse matrices; see 'else'\n",
    "    inv = 1.0/cell_area\n",
    "    D = sparse.diags(inv)\n",
    "    norm_counts = D.dot(raw_counts)\n",
    "    scaled_counts = norm_counts.multiply(median_area)\n",
    "    log_norm_counts = np.log1p(scaled_counts)\n",
    "else:\n",
    "    # simpler self explanatory code for dense matrices\n",
    "    scaled_counts = (raw_counts/cell_area[:,None])*median_area\n",
    "    log_norm_counts = np.log1p(scaled_counts)\n",
    "\n",
    "adata.layers['XENIUM_SIZE_NORM'] = log_norm_counts\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73734f07-27a0-4a33-b560-e41a846dc0ca",
   "metadata": {},
   "source": [
    "We can see that for some genes, there’s little difference - e.g. B-Cell marker MS4A1/CD20:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9972afdd-f8ab-4481-b246-d77c31630b38",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plot_feature = \"MS4A1\"\n",
    "\n",
    "fig, ax = plt.subplots(1, 2, figsize = (12, 5))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = plot_feature,\n",
    "    layer = \"XENIUM_SIZE_NORM\",\n",
    "    ax = ax[0],\n",
    "    show = False,\n",
    ")\n",
    "\n",
    "ax[0].set_title('Area Size - XENIUM_SIZE_NORM assay')\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = plot_feature,\n",
    "    layer = \"COUNT_NORM\",\n",
    "    ax = ax[1],\n",
    "    show = False,\n",
    ")\n",
    "\n",
    "ax[1].set_title(\"COUNT_NORM\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58c9eb04-9341-4d74-8231-bb9c03016474",
   "metadata": {},
   "source": [
    "But for others, standard normalisation and area size are very different:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e4e9ab-e340-44d3-b88d-387c69419105",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plot_feature = \"TUBB\"\n",
    "\n",
    "fig, ax = plt.subplots(1, 2, figsize = (12, 5))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = plot_feature,\n",
    "    layer = \"XENIUM_SIZE_NORM\",\n",
    "    ax = ax[0],\n",
    "    show = False,\n",
    ")\n",
    "\n",
    "ax[0].set_title('Area Size - XENIUM_SIZE_NORM assay')\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = plot_feature,\n",
    "    layer = \"COUNT_NORM\",\n",
    "    ax = ax[1],\n",
    "    show = False,\n",
    ")\n",
    "\n",
    "ax[1].set_title(\"COUNT_NORM\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea37f1c5-196a-4142-a736-5ca9b9f74a14",
   "metadata": {},
   "source": [
    "You can see that the most difference correlates with large stromal cells.\n",
    "But, we can probably agree that due to nuclei-expansion approach to cell segmentation in this dataset, the area size of these cells is also likely to be overly inflated, and therefore area-based normalisation here is probably strongly under-estimating the true expression and creating a false gene expression pattern!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba6437fb-0072-419f-ac16-a6205cd9f38e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = 'cell_area',\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71342413-d836-4de3-914f-0623bf11c5b8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Cell type identification\n",
    "\n",
    "You can manually annotate your cell clusters, or you can classify them using a reference single-cell dataset.\n",
    "This process is simpler than for Visium data because our data is at the single-cell level, establishing a one-to-one relationship without the need for spot deconvolution.\n",
    "\n",
    "However, our transcriptome is more limited here, and some cell types may not be well represented. Additionally, our single-cell reference might be missing some cell types that are not well captured by droplet-based technologies but are present in our tissue data.\n",
    "\n",
    "In this example, we will use a single-cell reference dataset that we prepared earlier.\n",
    "\n",
    "We will start by reading in the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "552b6647-c6a9-4ac6-ba43-8ff676d54ca7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "file_path = \"/nvme/project/shared/python/5_python_spatial_omics/2_imaging/data/training_course_ref.h5ad\"\n",
    "adata_ref = sc.read_h5ad(file_path)\n",
    "adata_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f8338fc-b583-4f15-bad9-75e25528115c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_ref.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "960f5e09-5fce-4930-84e3-e85ba4933e67",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata_ref,\n",
    "    color = \"CellType\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b925a95-ca8a-4344-abd1-1e66e2c867db",
   "metadata": {},
   "source": [
    "To regroup, we will also import an `adata` object from file, in case any of us diverged from each other in the worflow so far:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f25d5918-4d6c-4b92-9b49-7e13d11ba570",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "file_path = \"/nvme/project/shared/python/5_python_spatial_omics/2_imaging/data/training_course_adata_preprocessed_new.h5ad\"\n",
    "adata = sc.read_h5ad(file_path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa388fcd-f819-4680-9c23-4b0f54750faf",
   "metadata": {},
   "source": [
    "Before we proceed further, we subset both datasets to the set of genes they have in common:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f804557-3b45-4610-b36e-eb4c1e849d4a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "common_genes = adata.var_names.intersection(adata_ref.var_names)\n",
    "adata_ref = adata_ref[:, common_genes].copy()\n",
    "adata = adata[:, common_genes].copy()\n",
    "len(common_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4505e36a-9a73-478c-b11e-e1aef5148e58",
   "metadata": {},
   "source": [
    "We want to evaluate how much structural information is lost in single-cell data when limiting ourselves to the targeted gene set. Accurate cluster prediction is challenging if the current gene set does not adequately identify them. To do this, we will quickly re-embedd the data using only the genes present in our spatial transcriptomics data and keep the original cluster annotations derived from unbiased data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edfa8b45-e602-46c3-a9e0-69a49a663520",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.pca(\n",
    "    adata_ref, \n",
    "    n_comps=50, \n",
    "    use_highly_variable=False, \n",
    "    svd_solver='arpack', \n",
    "    copy=False\n",
    ")\n",
    "\n",
    "\n",
    "sc.pp.neighbors(\n",
    "    adata_ref, \n",
    "    n_neighbors=20, \n",
    "    use_rep='X_pca'\n",
    ")\n",
    "\n",
    "sc.tl.umap(adata_ref, min_dist=0.3)\n",
    "\n",
    "adata_ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20969f47-0c42-4728-962d-7e9d9dfffb47",
   "metadata": {},
   "source": [
    "In this example, we can observe that the limited gene set does a reasonably good job at distinguishing major cell populations. However, it struggles to differentiate between similar cell types, such as myofibroblasts and fibroblasts, as effectively as before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a41cacb4-ad6e-414d-9a12-e51263f8445c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.embedding(\n",
    "    adata_ref,\n",
    "    basis=\"X_umap\",\n",
    "    color=\"CellType\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94a7bbc5-bf2c-4041-9e15-2c19be377e8d",
   "metadata": {},
   "source": [
    "If we visualise the specificity of the gene panel across our single cell reference clusters, we can see that the panel coverage is mainly concentrated across epithelial cells and T-Cells and other immune cells, with few specific markers expressed by stromal cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a62373f0-77c0-4714-834b-337e1cc2cb29",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(16, 6))\n",
    "\n",
    "sc.pl.matrixplot(\n",
    "    adata_ref,\n",
    "    var_names=adata_ref.var_names,\n",
    "    groupby='CellType',\n",
    "    standard_scale=\"var\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b2c7a3-ca87-403b-8e00-5237cd51013a",
   "metadata": {},
   "source": [
    "A comparable heatmap can be obtained by aggregating data by cell type ourselves and using the `sc.pl.heatmap()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bae2e1e-3b31-4091-bf2d-de1013c0c81b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "aggregated = sc.get.aggregate(\n",
    "    adata_ref, by=\"CellType\", func=[\"mean\"]\n",
    ")\n",
    "aggregated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ebb51aa-63ee-4b11-87e7-e9646ed7a8c1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.heatmap(\n",
    "    aggregated,\n",
    "    layer=\"mean\",\n",
    "    var_names=aggregated.var_names,\n",
    "    groupby=\"CellType\",\n",
    "    dendrogram=False,\n",
    "    standard_scale=\"var\",\n",
    "    figsize=(16, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f1ac794-5c3b-421a-9931-4fd71d3fb75b",
   "metadata": {},
   "source": [
    "Next, we can use the standard `scanpy.tl.ingest()` integration and cross-classification workflow to transfer single-cell derived labels to our spatial object.\n",
    "\n",
    "Briefly, the function integrates embeddings and annotations of an adata with a reference dataset adata_ref through projecting on a PCA (or alternate model) that has been fitted on the reference data. The function uses a knn classifier for mapping labels and the UMAP package [McInnes et al., 2018] for mapping the embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5315d1e-14fd-4569-aa87-544990ac1ab4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.X = adata.X.toarray()\n",
    "sc.tl.ingest(adata, adata_ref, obs='CellType', embedding_method=['pca'])\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "462f6df1-6fb9-4b05-99f8-709b88008ab9",
   "metadata": {},
   "source": [
    "Unfortunately, the predicted labels and spatial clusters do not correspond clearly in all cases. This discrepancy is particularly evident in the middle regions of the UMAP, where many cells are predicted as epithelial cells - probably incorrectly!\n",
    "\n",
    "How to improve this?\n",
    "\n",
    "**Ensure Good Representation of Cell Type Markers in *in situ* Target Panel:**\n",
    "Most critically, before undertaking any experiments, you want to ensure that there is good representation of all cell types in your target panel.\n",
    "In this case, there is not much to be done as the data has already been generated.\n",
    "\n",
    "**Review and Refine Reference Data:**\n",
    "Ensure that the reference single-cell dataset is comprehensive and accurately annotated. If certain cell types are not well represented or annotated in the reference dataset, it can lead to misclassification.\n",
    "\n",
    "**Increase the Number of Dimensions:**\n",
    "Increasing the number of dimensions used in the UMAP and PCA steps might capture more variance in the data, leading to better label transfer.\n",
    "\n",
    "**Filter and Preprocess Data:**\n",
    "Filtering out low-quality cells or genes and performing additional preprocessing steps can enhance the accuracy of the transfer anchors and, consequently, the label predictions.\n",
    "\n",
    "**Manually Annotate or Correct Predictions:**\n",
    "In cases where automatic label transfer is insufficient, consider manually annotating or correcting the predictions for critical regions to ensure accuracy.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "582d555f-b1a8-4091-b187-8e3de3ebb902",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = \"CellType\",\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "277e283e-fd11-4dee-b527-a4a7de8c13dc",
   "metadata": {},
   "source": [
    "As before, we can also visualise the predicted cell labels in tissue space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ad90f73-33da-4133-980a-0ca452083dd8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(12, 6))\n",
    "\n",
    "sc.pl.spatial(\n",
    "    adata,\n",
    "    color=\"CellType\",\n",
    "    spot_size =10,\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc70e885-8741-4ab4-bf27-3ecbc7a38399",
   "metadata": {},
   "source": [
    "Including zoomed in views of regions of interest:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff9ffef2-4408-4513-a17d-a191d98e5c05",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=\"CellType\",\n",
    "    crop_coord=(4000, -9000, 5000, -8000),\n",
    "    figsize=(8, 6),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64ed8ce0-93e0-4b53-8a5b-ea8a22bd2f0d",
   "metadata": {},
   "source": [
    "Unfortunately, `ingest()` does not provide any confidence score on its prediction.\n",
    "\n",
    "When other methods provide such scores, it can be helpful to visualise and diagnose issues such as:\n",
    "\n",
    "- lower confidence in certain spatial areas\n",
    "- lower confidence in cells embed \"between\" clusters (e.g. between B-cells and T-cells)\n",
    "\n",
    "This is often the case where cell segmentation is imperfect and partitions transcripts in such a way that it generates “artificial” doublets by pulling in transcripts from an adjacent cell."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3de882e-dc74-4262-bbbe-21641529ee5d",
   "metadata": {
    "tags": []
   },
   "source": [
    "For example, if we visualise the lineage markers for T-Cells and B-Cells, we can see that they are often “co-expressed” in the same cells when biologically, they should not be.\n",
    "\n",
    "The `sc.pl.scatter()` function is used to create a scatter plot showing the relationship between the expression levels of two genes across all cells.\n",
    "This visualization helps to identify potential correlations or patterns between the two genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3655e406-ac46-411b-b2f7-d6dd6bafc44f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.scatter(\n",
    "    adata,\n",
    "    x=\"MS4A1\",\n",
    "    y=\"CD3D\",\n",
    "    color=\"leiden\",\n",
    "    size=20\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39a49c43-4c29-4743-918c-ce3653d27be6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = [\"MS4A1\", \"CD3D\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83429b76-e9f9-483f-b014-4164c4fd0015",
   "metadata": {},
   "source": [
    "To improve these artefacts, we can try alternative cell segmentation algorithms.\n",
    "What works best is very tissue dependant and there’s no easy one stop solution to this.\n",
    "Cell segmentation algorithms can be divided into a few groups.\n",
    "\n",
    "**Nuclei-based Segmentation** algorithms primarily focus on identifying cell nuclei, which are usually more distinct and easier to detect than the cell boundaries.\n",
    "Once the nuclei are identified, the cell boundaries are inferred by expanding around the nuclei.\n",
    "This approach works well in tissues where the nuclei are clearly visible and distinct and in early versions of many in situ platforms, were the only available methods due to only using DAPI stain.\n",
    "\n",
    "**Cell Boundary-Based Segmentation** algorithms (e.g. Cellpose) directly segments cells by identifying their boundaries.\n",
    "It is particularly effective for images with complex cell shapes and varying sizes, but this required good cell boundary staining - this is not available for our test dataset.\n",
    "Often cell boundary staining can be non-uniform across different tissues, adding further difficulties.\n",
    "Cellpose version 3 incorporates user-guided model training, which can be very useful for difficult to segment cell types - but this requires time investment to annotate training examples.\n",
    "\n",
    "**Transcript-Density Based Segmentation** algorithms, like Baysor segments cells based on the spatial distribution of transcripts\n",
    "It uses Bayesian inference to assign transcripts to cells, considering both the density and distribution of RNA molecules. This can be very useful for improving cell segmentation where cell boundary stain is not available or not working well.\n",
    "\n",
    "In this case, we will try re-segmenting our data with **Baysor**.\n",
    "Here’s the run we prepared earlier - see supplementary material on how to process the data yourself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5352a8d9-59b0-4957-b027-0b853961e4ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "baysor_dir = \"/nvme/project/shared/python/5_python_spatial_omics/2_imaging/data/baysor\"\n",
    "\n",
    "seg = pd.read_csv(\n",
    "    os.path.join(\n",
    "        baysor_dir, \"segmentation.csv\"\n",
    "    )\n",
    ")\n",
    "\n",
    "seg.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aa440c8-83c9-4ee1-accf-fd1a4ff2e189",
   "metadata": {},
   "source": [
    "There will be some transcripts that cannot be assigned to a cell - about 10% in this case.\n",
    "This information is stored under “is_noise” flag.\n",
    "This is fairly normal levels of noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f61dc70b-9446-4f29-8f4e-a3a16128973f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "seg['is_noise'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e880a01c-75fa-41e5-8ed6-d9339277cc51",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 5))\n",
    "plt.hist(x=seg['assignment_confidence'], bins=50)\n",
    "plt.xlabel('Seg Assignment Confidence')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "769547c0-bf5a-41ee-b2a5-a10a3b3d0794",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "(seg['assignment_confidence'] > 0.9).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae7e185-4b9d-4ba8-b338-121967b05440",
   "metadata": {},
   "source": [
    "plt.hist(x=seg['confidence'], bins=50)\n",
    "plt.xlabel('Seg Confidence')\n",
    "plt.show()And transcript confidence - the confidence that the molecule itself is real and not noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aed1023e-0b63-416b-8a97-87b8acc4f1e7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 5))\n",
    "plt.hist(x=seg['confidence'], bins=50)\n",
    "plt.xlabel('Seg Confidence')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3f357e8-8ef8-4197-82d8-5fcddaef5df0",
   "metadata": {},
   "source": [
    "We can filter out low confidence and low assignment confidence transcripts here from further analysis.\n",
    "How stringent you want to be depends on whether you want to keep as much data as possible and accept some inaccuracies, or end up with the cleanest possible dataset.\n",
    "\n",
    "Here, we will filter out transcripts that have not been assigned to cells, and below 0.9 confidence and assignment confidence.\n",
    "\n",
    "Then, we tabulate a cell by gene matrix from these data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eddcc5ed-3a45-4214-b6f4-6279352c0890",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "seg_filtered = seg[\n",
    "    (seg['confidence']> 0.9)&\n",
    "    (seg['assignment_confidence'] > 0.9) &\n",
    "    (~seg['is_noise'])\n",
    "]\n",
    "\n",
    "mat = pd.crosstab(seg_filtered['gene'], seg_filtered['cell'])\n",
    "mat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "059c1f08-a808-40bb-a0c7-d59f797391a3",
   "metadata": {},
   "source": [
    "Baysor further provides diagnostic info about cells in “segmentation_cell_stats.csv” file, which we will also read in here.\n",
    "\n",
    "The following parameters can be used to filter low-quality cells:\n",
    "\n",
    "**area:**\n",
    "area of the convex hull around the cell molecules avg_confidence: average confidence of the cell molecules\n",
    "\n",
    "**density:**\n",
    "the number of molecules in a cell divided by the cell area\n",
    "\n",
    "**elongation:**\n",
    "ratio of the two eigenvalues of the cell covariance matrix\n",
    "\n",
    "**n_transcripts:**\n",
    "number of molecules per cell\n",
    "\n",
    "**avg_assignment_confidence:**\n",
    "average assignment confidence per cell.\n",
    "Cells with low avg_assignment_confidence have a much higher chance of being an artifact.\n",
    "\n",
    "**max_cluster_frac (only if n-clusters > 1):**\n",
    "fraction of the molecules coming from the most popular cluster.\n",
    "Cells with low max_cluster_frac are often doublets.\n",
    "\n",
    "**lifespan:**\n",
    "number of iterations the given component exists.\n",
    "The maximal lifespan is clipped proportionally to the total number of iterations.\n",
    "Components with a short lifespan likely correspond to noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc27df0a-852b-425c-af0e-1e369399872f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "stats = pd.read_csv(\n",
    "    os.path.join(baysor_dir,\"segmentation_cell_stats.csv\")\n",
    ")\n",
    "stats.set_index('cell', inplace=True)\n",
    "\n",
    "stats.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4854077b-479d-4a16-bcfc-f80b9aac1a90",
   "metadata": {},
   "source": [
    "Now, we can assemble an AnnData object as before:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32e9e2e1-9673-44f6-95d8-68ba22cc7e31",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_reseg = AnnData(mat.values.T)\n",
    "adata_reseg.var_names = mat.index\n",
    "adata_reseg.obs_names = mat.columns\n",
    "adata_reseg.obs = stats.loc[mat.columns]\n",
    "\n",
    "centroids = adata_reseg.obs[['x', 'y']].to_numpy()[:,[1,0]]\n",
    "centroids[:,1] = -centroids[:, 1]\n",
    "\n",
    "adata_reseg.obsm['spatial'] = centroids\n",
    "\n",
    "adata_reseg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14f74f8c-7fd2-44ea-8b57-66d6fb6544ad",
   "metadata": {},
   "source": [
    "Different segmentation algorithms output cell boundaries in various cell formats.\n",
    "\n",
    "[GeoJSON](https://geojson.org/) is a format for encoding various geographic data structures using JavaScript Object Notation (JSON).\n",
    "It is widely used for representing spatial features and their attributes and is used by some algorithms to store and output cell segmentation boundaries.\n",
    "\n",
    "In Python, we can read in polygon data from a JSON file using the `json.load()` function, followed by additional processing using `geopandas`.\n",
    "\n",
    "*NOTE: baysor is a 3D cell segmentation algorithm.\n",
    "This means it considered z-stack information.\n",
    "Some algorithms only perform cell segmentation on a representative layer - e.g. Merscope Cellpose takes the middle z-stack.\n",
    "This resegmented data represents 3D segmentations of cells.\n",
    "Since these are 3D segmentations, they may look unusual when visualized as 2D projections.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7cfaed5-50e7-4fc5-8cb7-d9fce8a68bfe",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "polygons_path = os.path.join(\n",
    "    baysor_dir, \"segmentation_polygons.json\"   \n",
    ")\n",
    "\n",
    "with open(polygons_path, 'r') as f:\n",
    "    geo = json.load(f)\n",
    "\n",
    "\n",
    "geo.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af2a361f-3900-44f2-be78-fe62679aae00",
   "metadata": {},
   "source": [
    "In the below code, we extract the polygon coordinates from the data and reformat them into a data frame that `geopandas` requires."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f05e372f-545b-4f27-8778-89840e891337",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from shapely import Polygon\n",
    "import geopandas as gpd\n",
    "\n",
    "poly_geoDF = []\n",
    "\n",
    "for i,g in enumerate(geo[\"geometries\"]):\n",
    "    cells_id = g['cell']\n",
    "    if cells_id is None:\n",
    "        continue\n",
    "    # print(g['coordinates'][0])\n",
    "    if len(g[\"coordinates\"][0])> 2:\n",
    "        cell_polygon = Polygon(g[\"coordinates\"][0])\n",
    "        poly_geoDF.append({'cell_id': cells_id, 'geometry': cell_polygon})\n",
    "\n",
    "gdf = gpd.GeoDataFrame(poly_geoDF).set_index(\"cell_id\")\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b856dfe9-1352-4987-a9ca-641b5ee535b7",
   "metadata": {},
   "source": [
    "From here, we can use the seurat object to visualise various cell meta data - for example, average transcript assignment confidence per cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6875ccf-4b91-4dc4-aa62-a80f372b2895",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = sq.pl.spatial_scatter(\n",
    "    adata_reseg,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    size=2,\n",
    "    color=\"avg_assignment_confidence\",\n",
    "    figsize=(12, 6),\n",
    "    return_ax=True,\n",
    ")\n",
    "ax.set_facecolor(\"black\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13152eaa-055a-4a6c-b776-1e6533764cf1",
   "metadata": {},
   "source": [
    "Let's filter out low count cells and re-cluster the data as before.\n",
    "In practice, you may also want to consider additional QC metrics here - e.g. removing overly small cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e399c2be-9e7a-43e9-bdb2-0d1df7bce4cc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(adata_reseg, inplace=True, log1p=False)\n",
    "\n",
    "adata_reseg = adata_reseg[\n",
    "    adata_reseg.obs[\"total_counts\"] >= 15\n",
    "].copy()\n",
    "\n",
    "sc.pp.normalize_total(adata_reseg, target_sum=1e4)\n",
    "sc.pp.log1p(adata_reseg)\n",
    "sc.pp.scale(adata_reseg)\n",
    "\n",
    "sc.pp.pca(\n",
    "    adata_reseg, \n",
    "    n_comps=25, \n",
    "    use_highly_variable=False, \n",
    "    svd_solver='arpack', \n",
    "    copy=False\n",
    ")\n",
    "\n",
    "sc.pp.neighbors(\n",
    "    adata_reseg, \n",
    "    n_neighbors=20, \n",
    "    use_rep='X_pca'\n",
    ")\n",
    "\n",
    "sc.tl.umap(adata_reseg, min_dist=0.3)\n",
    "\n",
    "sc.tl.leiden(\n",
    "    adata_reseg,\n",
    "    resolution = 0.5\n",
    ")\n",
    "\n",
    "adata_reseg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5108b15e-e938-4200-b95d-d06bdf476994",
   "metadata": {},
   "source": [
    "Visualising clusters, we can see that we already obtain a better separation in the UMAP embedding than before.\n",
    "Though of course, distances in the UMAP space can be very misleading and careful interpretation is required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b856c54e-cba6-4c81-a773-f7d4cc936224",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata_reseg,\n",
    "    color = 'leiden',\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ae40cb-279e-4132-9db6-de4b1a98cc56",
   "metadata": {},
   "source": [
    "Next we visualise the clusters in tissue space.As before, lets cross-classify our cells using the reference single cell dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c197066f-9f6f-482e-81a6-77ebe834a6e4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(12, 6))\n",
    "\n",
    "sc.pl.spatial(\n",
    "    adata_reseg,\n",
    "    color = 'leiden',\n",
    "    spot_size=10,\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ebc9857-c355-461b-9339-4d6b4f4ccaaf",
   "metadata": {},
   "source": [
    "As before, lets cross-classify our cells using the reference single cell dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4bf76a2-f23a-4602-8ee5-cb8bc4f6c517",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "file_path = \"/nvme/project/shared/python/5_python_spatial_omics/2_imaging/data/training_course_ref.h5ad\"\n",
    "adata_ref = sc.read_h5ad(file_path)\n",
    "\n",
    "common_genes = adata_reseg.var_names.intersection(adata_ref.var_names)\n",
    "adata_ref = adata_ref[:, common_genes].copy()\n",
    "adata_reseg = adata_reseg[:, common_genes].copy()\n",
    "\n",
    "sc.pp.pca(\n",
    "    adata_ref, \n",
    "    n_comps=25, \n",
    "    use_highly_variable=False, \n",
    "    svd_solver='arpack', \n",
    "    copy=False\n",
    ")\n",
    "\n",
    "\n",
    "sc.pp.neighbors(\n",
    "    adata_ref, \n",
    "    n_neighbors=20, \n",
    "    use_rep='X_pca'\n",
    ")\n",
    "\n",
    "sc.tl.umap(adata_ref, min_dist=0.3)\n",
    "\n",
    "sc.tl.ingest(adata_reseg, adata_ref, obs='CellType', embedding_method=['pca'])\n",
    "\n",
    "adata_reseg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d619f19e-6886-4403-97ac-9f87dc3eab72",
   "metadata": {},
   "source": [
    "Visualising the predictions, we’ve separated T-Cells from B-Cells much better.\n",
    "The stromal clusters still predict poorly, but that is due to poor probe coverage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b362028-bc0f-4d21-853b-e12a16032a9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata_reseg,\n",
    "    color = 'CellType',\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "147165e2-d18c-4198-911a-39ada4568e35",
   "metadata": {},
   "source": [
    "We can check the distribution in tissue space:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be451cea-97ae-420c-8619-e1b7b3fcc0d2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "sc.pl.spatial(\n",
    "    adata_reseg,\n",
    "    color = \"CellType\",\n",
    "    spot_size=10,\n",
    "    ax=ax\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eee5de0-74ac-4eb5-8989-c7959b08012d",
   "metadata": {},
   "source": [
    "### Additional Exercises\n",
    "\n",
    "1. How does this compare with other segmentation algorithms? For example, proseg - data resegmented with proseg has been made available here: /nvme/project/shared/r/3_r_spatial/DATA/PRECOMPUTED/proseg\n",
    "2. What metrics would you use to quantify cell segmentation accuracy?\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
