{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "67a545a6-86ef-4118-b6be-cc980a4b45a5",
   "metadata": {},
   "source": [
    "# DAY 3: Imaging-based Spatial Transcriptomics Analysis - Xenium Human Colon Dataset\n",
    "\n",
    "This notebook will guide you through Day 1 content. We will cover:\n",
    "\n",
    "1. Setting up the environment\n",
    "2. Loading the data\n",
    "3. Spatial visualisation\n",
    "4. Quality control\n",
    "5. Data normalisation, Dimensionality reduction, and Clustering\n",
    "6. Cell size normalisation\n",
    "7. Cell type label prediction using single cell references\n",
    "8. Improving cell segmentation\n",
    "\n",
    "## How to use this notebook\n",
    "\n",
    "This notebook is intended to be used as a reference for your own analysis.\n",
    "All code chunks have an explanation detailing the analysis steps and their purpose, as well as key parameters.\n",
    "Play around with these and see what they do so you are better equiped to adapt the workflow to your own data.\n",
    "\n",
    "## Dataset\n",
    "\n",
    "We will be using a Xenium human colon dataset, generated using V1 Xenium chemistry with a small, off the shelf premade colon target gene panel and nuclei staining (no cell-boundary staining).\n",
    "The full dataset is available here and contains one large tissue section from human colon and one large colorectal tissue section.\n",
    "Typically, a full Xenium slide can generate data from over a million cells.\n",
    "Therefore, several basic analysis steps can take a while to run.\n",
    "For the purposes of this tutorial, we split off a small field of view from the full dataset.\n",
    "If you're interested in trying working with the full size dataset, you can find it in the data folder.\n",
    "\n",
    "More information about today's dataset can be found on the 10x website.\n",
    "\n",
    "https://www.10xgenomics.com/datasets/human-colon-preview-data-xenium-human-colon-gene-expression-panel-1-standard"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4128c4b7-f228-49ba-9aa5-98ba1c5b5b79",
   "metadata": {},
   "source": [
    "### 1. Set up environment\n",
    "\n",
    "First we need to set up the environment and load the packages we will use for this workshop.\n",
    "As in the sequencing-based spatial transcriptomics analysis tutorial, we will be using largely the same base packages for analysis - [scanpy](https://scanpy.readthedocs.io/en/stable/) and [squidpy](https://squidpy.readthedocs.io/en/stable/). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33da7aef-01f2-4d2b-9a15-3b1a2ba9a820",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.io import mmread\n",
    "from anndata import AnnData\n",
    "import squidpy as sq\n",
    "from spatialdata_io import xenium\n",
    "import spatialdata as sd\n",
    "import spatialdata_plot \n",
    "from spatialdata import SpatialData\n",
    "from scipy import sparse\n",
    "from scipy.cluster.hierarchy import linkage, leaves_list\n",
    "from scipy.spatial.distance import pdist\n",
    "import celltypist\n",
    "from celltypist import models"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58962e23-8657-401d-be31-3fb6c01952c0",
   "metadata": {},
   "source": [
    "Lets set the paths to where the datasets are stored and where we will keep any outputs we calculate ourselves. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4841334-9494-4b7a-b895-c8167ced94ee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Input data\n",
    "DATA_DIR = \"/nvme/project/shared/python/5_python_spatial_omics/data/xenium_human_colon\"\n",
    "# Change this to where you want to save your outputs\n",
    "OUTPUT_DIR = \"/PATH/TO/YOUR/DIRECTORY\"\n",
    "# Path to precomputed data that you can use to skip lengthy steps\n",
    "PRECOMPUTED_DIR = \"/nvme/project/shared/python/5_python_spatial_omics/data/precomputed/imaging/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15733cbe-ad42-4ac8-b7d8-5d6b5087dc42",
   "metadata": {},
   "source": [
    "### 2. Loading the data\n",
    "\n",
    "Typically, we can load xenium data using spatial data package. Spatial data provides a nice data structure for xenium data where we can load cell segmentation shapes, gene expression data, individual transcript coordinates and also any images. \n",
    "\n",
    "In this case, as this is Xenium V1 experiment before cell boundary staining was released, we only have a DAPI image, which is not very exciting or informative, so we skip loading that for now. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81e32889-4ff6-45ba-a848-9bd8dc9100b5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen = xenium(\n",
    "    DATA_DIR,\n",
    "    transcripts=True,\n",
    "    aligned_images=False,\n",
    "    morphology_mip=True,\n",
    "    morphology_focus=True, \n",
    "    cells_as_circles=False, \n",
    "    nucleus_boundaries = False,\n",
    "    n_jobs=20\n",
    ")\n",
    "xen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f63ab1-462a-4a35-86a3-3441c8b99146",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen.write(\n",
    "    os.path.join(OUTPUT_DIR, \"xen_full.zarr\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da720f1d-527f-4335-90a4-06e4e7b71b23",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen = SpatialData.read(\n",
    "    os.path.join(OUTPUT_DIR, \"xen_full.zarr\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6deb92c8-c620-418b-8807-a2bcac6e4bb3",
   "metadata": {},
   "source": [
    "As the whole slide is quite a large piece of tissue with many cells, for the sake of this tutorial we will be working with a subset of the data.\n",
    "Here, we subset the original spatial data object to a bounding box containing a smaller number of cells. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61df6723-d758-4536-a14d-4f41660088eb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped = xen.query.bounding_box(\n",
    "    axes=[\"x\", \"y\"],\n",
    "    min_coordinate=[35000, 10000],\n",
    "    max_coordinate= [45000, 25000],\n",
    "    target_coordinate_system=\"global\"\n",
    ")\n",
    "xen_cropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3e4137b-ebf6-45b1-b94b-aa191717968f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a00add3-1d96-4a56-bd08-fc24c4c2cad2",
   "metadata": {},
   "source": [
    "We can save this subset to disk or reload it. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20982700-1bcd-4c71-9c7f-c848c7bcae71",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped.write(\n",
    "    os.path.join(OUTPUT_DIR, \"xen_cropped.zarr\"),\n",
    "    overwrite=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aeb4278c-39e9-40a3-81aa-1ac31fc20ebe",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped = SpatialData.read(\n",
    "    os.path.join(OUTPUT_DIR, \"xen_cropped.zarr\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35b704b4-533c-4a0c-a1a8-0c04c5606b23",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e0e238c-466f-48f8-974d-65b8b42821af",
   "metadata": {},
   "source": [
    "Spatial data objects have an anndata object stored in the table slot, which we can examine and operate on by accessing it directly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe9624e-d4f4-481b-8856-df52baabe66e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d477e30-f183-4558-bcd2-60849ea71bb1",
   "metadata": {},
   "source": [
    "### 3. Spatial Visualisation\n",
    "\n",
    "Let’s start with some basic QC and visualisation of the data.\n",
    "\n",
    "Spatial data has its own set of plotting functions, which are documented here:\n",
    "https://spatialdata.scverse.org/projects/plot/en/latest/plotting.html\n",
    "\n",
    "In summary:\n",
    "\n",
    "**render_labels:**\n",
    "We can use this to plot cell centroids.\n",
    "This should be your default visualisation, especially for larger datasets as plotting is much quicker.\n",
    "\n",
    "**render_images:**\n",
    "We can plot image data, if there is any linked to the object.\n",
    "In our case, there is not.\n",
    "\n",
    "**render_points:**\n",
    "We can use this to plot individual transcript coordinates.\n",
    "\n",
    "**render_shapes:**\n",
    "We can use this to plot cell segmentation boundaries - but this can be slow for large datasets. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a0faa6-488b-4f9b-9bb8-63b8da259863",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Cell centroids\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6092c5ee-d32c-4b63-b72e-f8ae0a210b66",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped.pl.render_images(\n",
    "    \"morphology_focus\", cmap=\"gray\"\n",
    ").pl.show(\n",
    "    title=\"Morphology image\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba9cd895-feb7-466a-84da-f4efaa9def17",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped.tables[\"table\"].obs[\"region\"] = \"cell_boundaries\"\n",
    "\n",
    "xen_cropped.set_table_annotates_spatialelement(\n",
    "    table_name=\"table\",\n",
    "    region=\"cell_boundaries\",\n",
    "    region_key=\"region\",\n",
    "    instance_key=\"cell_id\",\n",
    ")\n",
    "\n",
    "xen_cropped.pl.render_shapes(\n",
    "    \"cell_boundaries\",\n",
    "    table_name=\"table\",\n",
    "    color=\"total_counts\",\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Total transcript counts per cell + boundaries\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acded5b3-943f-4488-a572-1f4d0ef91510",
   "metadata": {},
   "source": [
    "We can't really easily visualise cell boundaries and shapes when we are zoomed out.\n",
    "Let's create a small zoomed-in ROI to see how we can visualise cell shapes.\n",
    "Here, we select a region focused on a lymphoid follicle in the tissue.\n",
    "You can also use colour and outline arguments to customise the look of the plot - here, we switched from default colour palette and added outlines to cell boundary shapes.\n",
    "We can also turn off axes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e653a5ac-fd51-4b79-bd5c-178c8e416865",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "zoom_roi = xen_cropped.query.bounding_box(\n",
    "    axes=[\"x\", \"y\"],\n",
    "    min_coordinate=[38000, 21000],\n",
    "    max_coordinate=[41000, 24000],\n",
    "    target_coordinate_system=\"global\",\n",
    ")\n",
    "\n",
    "ax = zoom_roi.pl.render_shapes(\n",
    "    \"cell_boundaries\",\n",
    "    table_name=\"table\",\n",
    "    color=\"total_counts\",\n",
    "    outline_color=\"black\",\n",
    "    outline_alpha=1,\n",
    "    outline_width=0.5,\n",
    "    cmap=\"magma\"\n",
    "\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Total transcript counts per cell + boundaries\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f027d42-f25b-4bbc-bbc6-4a917b43534e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped.pl.render_points(\n",
    "    \"transcripts\",\n",
    "    color=\"feature_name\",\n",
    "    groups=[\"MKI67\", \"MS4A1\"],\n",
    "    palette=[\"red\", \"black\"]\n",
    ").pl.show(\n",
    "    title=\"MKI67 transcripts\",\n",
    "    coordinate_systems=\"global\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff1243d-d9f4-4bcd-949e-ce2240b71b8f",
   "metadata": {},
   "source": [
    "We can also combine different types of plots - for example, we can use cell boundaries as a base and overlay individual transcript points over the image.\n",
    "Here, we do it in the zoomed-in ROI over a follicle. \n",
    "\n",
    "Plotting individual transcripts can be useful as it allows us to see detection outside of segmented cell boundary areas.\n",
    "This can be useful if, for example, you have custom probes for pathogens, which will not segment out as cells, or to diagnose any segmentation or quality issues with your slide.\n",
    "For example, here we can see that there is some detection of *MS4A1*/*CD20* transcript outside the cells. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8609eca4-c7ca-41c6-990a-bc35d6f6ccfb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = zoom_roi.pl.render_shapes(\n",
    "    \"cell_boundaries\",\n",
    "    table_name=\"table\",\n",
    "    color=\"total_counts\",\n",
    "    outline_color=\"black\",\n",
    "    outline_alpha=1,\n",
    "    outline_width=0.5,\n",
    "    cmap=\"magma\"\n",
    "\n",
    ").pl.render_points(\n",
    "    \"transcripts\",\n",
    "    color=\"feature_name\",\n",
    "    groups=[\"MS4A1\"],\n",
    "    palette=[\"red\"],\n",
    "    size=0.5\n",
    ").pl.show(coordinate_systems=\"global\", title=\"Total transcript counts per cell + boundaries\", return_ax=True)\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "937f4041-e499-4da7-a79f-a3a0d330c5ff",
   "metadata": {},
   "source": [
    "While plotting cell boundaries can look nice, for larger tissue sections it can take a long time to generate each plot.\n",
    "Therefore, for interactive sessions we recommend using cell centroids for quicker plotting.\n",
    "Let's link our [AnnData](https://anndata.readthedocs.io/en/stable/) back to cell centroids for that. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ec807b8-8706-4fd9-a2da-64037143a86c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped.tables[\"table\"].obs[\"region\"] = \"cell_labels\"\n",
    "\n",
    "xen_cropped.set_table_annotates_spatialelement(\n",
    "    table_name=\"table\",\n",
    "    region=\"cell_labels\",\n",
    "    region_key=\"region\",\n",
    "    instance_key=\"cell_labels\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c368b1d-e24a-4e64-99a7-c78280dc98f2",
   "metadata": {},
   "source": [
    "### 4. Quality control\n",
    "\n",
    "Next, we want to do some basic QC.\n",
    "We can get a lot of information from the Xenium Ranger run summary document, but ultimately we still want to do cell-level QC of our dataset. \n",
    "\n",
    "In our [SpatialData](https://spatialdata.scverse.org/en/stable/) object, cell level data is stored in the [AnnData](https://anndata.readthedocs.io/en/stable/) table, which we can access via `[\"table\"]`.\n",
    "We can operate on this as on any [AnnData](https://anndata.readthedocs.io/en/stable/) object.\n",
    "For example, let's use the [scanpy](https://scanpy.readthedocs.io/en/stable/) helper function to calculate automated QC metrics as we did before.\n",
    "Because we are using a small, targetted gene panel, we skip the typical mitochondrial/ribosomal gene metrics, as we don't have any of these genes in the panel anyway. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fddfcb6b-443b-46cd-907f-979202ee87ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(\n",
    "    xen_cropped[\"table\"],  inplace=True, percent_top=[10, 20], log1p=True\n",
    ")\n",
    "xen_cropped[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62218673-6211-48fc-8718-79e818c5cb59",
   "metadata": {},
   "source": [
    "First, let's visualise the total number of transcripts detected per cell.\n",
    "\n",
    "As in scRNA-Seq data, this is the most basic measure of overall signal and how well the data looks.\n",
    "\n",
    "Unlike in scRNA-Seq data or unbiased sequencing-based ST, these measures are also very heavily dependent not only on the total RNA quantity of each cell and tissue quality, but also on the target panel used for the experiment.\n",
    "Under-represented cell types will naturally yield fewer transcripts.\n",
    "Finally, the quality of cell segmentation also plays a role.\n",
    "\n",
    "In this case, we can see that there are areas with higher and lower total transcripts detected, but there's a few cells which have very high counts.\n",
    "\n",
    "Understanding your tissue and target panel here is important to delineate where these differences are biological and where they may be technical."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc3fa673-cc49-4bb4-91e5-90089cf6a191",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"transcript_counts\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(coordinate_systems=\"global\", title=\"Transcripts per cell\", return_ax=True)\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01436c8c-79dc-48d9-bffe-128879a3c033",
   "metadata": {},
   "source": [
    "It can also help to visualise the distribution to see where the majority of your cells lie. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56d90a62-1deb-47da-ad3e-e6f49fabfeee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.kdeplot(\n",
    "    data=xen_cropped[\"table\"].obs,\n",
    "    x=\"transcript_counts\",\n",
    "    fill=True\n",
    ")\n",
    "plt.xlabel(\"Transcript Count Per Cell\")\n",
    "plt.ylabel(\"Density\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9b11ee3-271a-4919-861e-95eb58f06bb6",
   "metadata": {},
   "source": [
    "Similarly, we can visualise the total number of gene detected per cell.\n",
    "You can see that this is a bit less variable across tissue.\n",
    "\n",
    "This can also suggest that there cells at the top of the epithelial crypts in this sample with more genes detected than the rest of the tissue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98df7aec-3fd8-4e47-9017-4dd2fc4e9158",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"n_genes_by_counts\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Genes per cell\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9661e94d-ebb2-4064-b535-f56efb5fd61b",
   "metadata": {},
   "source": [
    "This code examines the distribution of the number of genes detected per cell in the [AnnData](https://anndata.readthedocs.io/en/stable/) object\n",
    "using a density plot.\n",
    "\n",
    "This is important for understanding the variability and distribution of detected features,\n",
    "which can help identify potential issues such as low-quality cells and determine any filtering thresholds\n",
    "that may need to be applied.\n",
    "\n",
    "If you’re coming from scRNA-Seq work, these low numbers probably look very alarming.\n",
    "How can you possibly work with 31 median genes per cell?\n",
    "\n",
    "Unlike scRNA-Seq data and sequencing-based ST, both gene dropouts and noise are much, much lower in in situ ST data. So, this is surprisingly perfectly fine for cell identification and clustering.\n",
    "\n",
    "We are also working with 100-fold fewer targetted genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a31dc3c-fb05-4897-b6d0-db52f1684939",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.kdeplot(\n",
    "    data=xen_cropped[\"table\"].obs,\n",
    "    x=\"n_genes_by_counts\",\n",
    "    fill=True\n",
    ")\n",
    "plt.xlabel(\"Gene Count Per Cell\")\n",
    "plt.ylabel(\"Density\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c39d366b-7fac-4056-a7d4-00721682bd9d",
   "metadata": {},
   "source": [
    "Next, we visualise cell area/size, which allows us to examine the spatial organization and potential heterogeneity of cell sizes within your tissue sample.\n",
    "\n",
    "Why do we get such a difference in spatial distribution of cell sizes?\n",
    "\n",
    "This could be due to biological differences between small and large cells - e.g. small cells like T-cells.\n",
    "\n",
    "However, here the signal correlates with areas of low cellularisation.\n",
    "Therefore, it is likely this is an artefact of nuclei expansion in cell segmentation.\n",
    "\n",
    "**What is Nuclei Expansion?**\n",
    "\n",
    "Nuclei expansion in cell segmentation refers to the process of enlarging the segmented nuclei regions to approximate the boundaries of the entire cells.\n",
    "This technique is used to better represent the actual cell boundaries when only the nuclei have been explicitly segmented/we only have DAPI and no additional cell boundary staining.\n",
    "The primary goal is to provide a more accurate estimation of the cellular area, which is crucial for various downstream analyses in spatial transcriptomics and single-cell studies.\n",
    "In this case, nuclei expansion is constrained either by maximum distance or other nearby cells - so, where there are no other nearby cells to “bump into”, the expansion generates artificially bigger cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8da490d4-64f0-40be-93c5-7b148c2f14e5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"cell_area\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Cell size\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "818882f2-9b77-4f74-8d17-4ea72de1689c",
   "metadata": {},
   "source": [
    "We can see that nucleus size is much more uniform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "478f927b-b38c-47ab-a8cb-e4a5a94130b0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"nucleus_area\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Nucleus size\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c99fbc1-8ab4-4917-acd7-20e441b71a86",
   "metadata": {},
   "source": [
    "We can further check that this is likely the case by plotting the ratio between nuclei and total cell area.\n",
    "We can see that there is a very big decrease in percentage of cell area occupied by nucleus in areas of low cell density.\n",
    "\n",
    "The cell-to-nucleus area ratio can also potentially provide insights into cell morphology, cell type and potential changes in cellular states or conditions.\n",
    "For example, T-Cells can often be quite well identified by this variable alone, as they have a small cytoplasm volume.\n",
    "However, without a cell boundary stain, this metric mainly captures segmentation artefacts, so be careful about over-interpretation!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fe89109-1854-4c84-bedc-ba98b80671e6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].obs[\"cell_nucleus_ratio\"] = xen_cropped[\"table\"].obs[\"nucleus_area\"] / xen_cropped[\"table\"].obs[\"cell_area\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56a8507b-95ee-4f06-b498-050803d63fcd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"cell_nucleus_ratio\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Nucleus-Cell Ratio\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "741b15db-8229-400d-816a-6a9422b4d7d9",
   "metadata": {},
   "source": [
    "In this case, we can see that as expected, there is generally a correlation between cell area and transcript detection rate.\n",
    "\n",
    "However, we also have a group of cells where this is not the case - very large cells but relatively few transcripts.\n",
    "These cells are mainly submucosal stromal cells which are very poorly covered by the panel 10x have used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb70575-981c-4cb3-af6d-5e469892549f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.scatterplot(\n",
    "    data=xen_cropped[\"table\"].obs,\n",
    "    x=\"total_counts\",\n",
    "    y=\"cell_area\",\n",
    "    s=5\n",
    ")\n",
    "plt.xlabel(\"Transcript count\")\n",
    "plt.ylabel(\"Cell Size\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70eefb08-6bf4-4c7b-824a-1537d638e17c",
   "metadata": {},
   "source": [
    "We can create a filter to remove the overly large cells from the analysis.\n",
    "Here, we can flag cells which are above 99th percentile in cell area.\n",
    "We can see this would remove a lot of submucosal cells.\n",
    "We might want to reconsider whether to apply this filter or not!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20c7c186-8412-4d29-9bb4-d659793ea6c8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].obs[\"size_filter_large\"] = xen_cropped[\"table\"].obs[\"cell_area\"] < np.quantile(xen_cropped[\"table\"].obs[\"cell_area\"], 0.99)\n",
    "xen_cropped[\"table\"].obs[\"size_filter_large\"] = xen_cropped[\"table\"].obs[\"size_filter_large\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4745f6f1-c7e8-49d8-8e8f-80cc7a5b9347",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"size_filter_large\",\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Large Cell Filter\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c3e394c-6ad7-40ad-b56a-a85b97ef5add",
   "metadata": {},
   "source": [
    "Similarly, we can flag small cells for filtering - very small cells are often segmentation artefacts.\n",
    "But, as we can see, this disproportionalely removes cells within the densely packed immune region. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed84e07b-60f9-4905-8968-e485da167739",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].obs[\"size_filter_small\"] = xen_cropped[\"table\"].obs[\"cell_area\"] > np.quantile(xen_cropped[\"table\"].obs[\"cell_area\"], 0.01)\n",
    "xen_cropped[\"table\"].obs[\"size_filter_small\"] = xen_cropped[\"table\"].obs[\"size_filter_small\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9ad992a-94f2-4844-90a2-a5a6e1da8280",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"size_filter_small\",\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Small Cell Filter\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8990a17b-2617-4317-bdbb-b532379781b8",
   "metadata": {},
   "source": [
    "We can check how these values correlate with gene detection rate.\n",
    "\n",
    "If we filter out small cells, we will remove cells with low numbers of genes detected.\n",
    "\n",
    "If we filter out large cells, this is not that biased towards overly large counts, as we saw before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2d18bdc-955b-452f-86c3-1dbd7b092416",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.violinplot(\n",
    "    data=xen_cropped[\"table\"].obs,\n",
    "    x=\"size_filter_small\",\n",
    "    y=\"n_genes_by_counts\",\n",
    "    palette=\"tab10\"\n",
    ")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36b4837a-0739-458c-8488-61b3a01397e0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sns.violinplot(\n",
    "    data=xen_cropped[\"table\"].obs,\n",
    "    x=\"size_filter_large\",\n",
    "    y=\"n_genes_by_counts\",\n",
    "    palette=\"tab10\"\n",
    ")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0461bfc-9e5d-4dc3-b90f-1da57cbdef3f",
   "metadata": {},
   "source": [
    "Adjusting the threshold for what is considered a “small cell” can have significant implications for your analysis, especially in areas with specific cell types such as T-cells, which are small and densely packed in follicular regions.\n",
    "This example demonstrates how changing the threshold to the 10th percentile affects the filtering.\n",
    "In this case, we would probably filter out a lot of good cells that we don’t want to lose!\n",
    "So, be careful when looking at these types of QC metrics!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e853bc58-3c40-4cb9-ab7c-a873a0689f60",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].obs[\"size_filter_small\"] = xen_cropped[\"table\"].obs[\"cell_area\"] > np.quantile(xen_cropped[\"table\"].obs[\"cell_area\"], 0.1)\n",
    "xen_cropped[\"table\"].obs[\"size_filter_small\"] = xen_cropped[\"table\"].obs[\"size_filter_small\"].astype(\"category\")\n",
    "\n",
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"size_filter_small\",\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Small Cell Filter\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67c33eb3-daff-4aec-b075-b614dc765f0b",
   "metadata": {},
   "source": [
    "Lets set this back to the original 1% threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4824f5b-5808-454b-b0ac-986df92be38b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].obs[\"size_filter_small\"] = xen_cropped[\"table\"].obs[\"cell_area\"] > np.quantile(xen_cropped[\"table\"].obs[\"cell_area\"], 0.01)\n",
    "xen_cropped[\"table\"].obs[\"size_filter_small\"] = xen_cropped[\"table\"].obs[\"size_filter_small\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e611ad-d451-444e-8d03-179c803535b6",
   "metadata": {},
   "source": [
    "The most important filter is the overall transcript detection.\n",
    "Empty cells or cells with very low transcript count cannot be taken forward for clustering analysis and it is extremely difficult to identify what they may be.\n",
    "Here, we set a threshold of minimum 15 transcripts.\n",
    "This seems quite low - for data from *in situ* platforms with low noise (Xenium, Merfish, Merscope), this is generally enough to cluster and identify cell types.\n",
    "If your data has more noise (e.g. CosMx), a higher threshold is more appropriate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa646b67-37bb-4ee6-95eb-d6c55da187d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].obs[\"transcript_filter\"] = xen_cropped[\"table\"].obs[\"total_counts\"] >= 15\n",
    "xen_cropped[\"table\"].obs[\"transcript_filter\"] = xen_cropped[\"table\"].obs[\"transcript_filter\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11086c7c-7d9e-42ff-b939-2ec03608821c",
   "metadata": {},
   "source": [
    "And we can visualise the cells that we would lose.\n",
    "\n",
    "We see that we disproportionately would filter out more cells from some regions than others.\n",
    "As pointed out previously, this is likely due to a combination of gene panel coverage in some regions and very small cells in densely packed regions like follicles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eec4888-c40c-4f3c-8d07-8b4437552050",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"transcript_filter\",\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Transcript Filter\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d464cab7-5789-4798-ba7f-5f5c72b5314d",
   "metadata": {},
   "source": [
    "Finally, visualizing the counts of negative control codewords, negative control probes, and unassigned codewords helps identify and understand technical artifacts and background noise in your spatial transcriptomics data.\n",
    "\n",
    "Here, we can see that all control probes and codewords produce yield very little signal, suggesting our data is good quality!\n",
    "\n",
    "In some cases, high amount of autoflourescence is the cells/tissue can sometimes generate false positive signal and this should be filtered out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efec20a3-995a-47a3-a1ff-bc34b2fe7b5b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"deprecated_codeword_counts\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Deprecated Codeword Counts\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a2000f9-920d-47d1-a968-b1ed6f7aa698",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"unassigned_codeword_counts\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Unassigned Codeword Counts\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10a5a38b-60ae-4e01-ade8-b5bd5f98c9fa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"control_codeword_counts\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Control Codeword Counts\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "529e67d5-d5d8-4554-abae-005c48c0afd9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"control_probe_counts\",\n",
    "    cmap=\"jet\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Control Probe Counts\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd22d135-5160-4b12-93e8-accab1c15412",
   "metadata": {},
   "source": [
    "Although the negative control signal is low, we can nonetheless create a filter to remove cells which have any, although in this case it is probably unnecessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f4e6490-c3b0-4dd9-a091-12ffe0b788c0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].obs[\"probe_filter\"] = (\n",
    "    (xen_cropped[\"table\"].obs[\"control_probe_counts\"] == 0) &\n",
    "    (xen_cropped[\"table\"].obs[\"control_codeword_counts\"] == 0) &\n",
    "    (xen_cropped[\"table\"].obs[\"unassigned_codeword_counts\"] == 0) & \n",
    "    (xen_cropped[\"table\"].obs[\"deprecated_codeword_counts\"] == 0)\n",
    ")\n",
    "\n",
    "xen_cropped[\"table\"].obs[\"probe_filter\"] = xen_cropped[\"table\"].obs[\"probe_filter\"].astype(\"category\")\n",
    "\n",
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"probe_filter\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Probe Filter\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a532bd17-7080-415a-9881-4c83aa3df575",
   "metadata": {},
   "source": [
    "Finally, we can subset the [AnnData](https://anndata.readthedocs.io/en/stable/) object based on any/all of the filters we have created earlier.\n",
    "\n",
    "By combining probe, size, and transcript filters, you can retain only the cells that meet all quality criteria, reducing the impact of technical artifacts and noise on your analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26d93665-6459-49d3-8a73-794bd9331482",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mask = (\n",
    "    (xen_cropped[\"table\"].obs[\"probe_filter\"].astype(bool)) &\n",
    "    (xen_cropped[\"table\"].obs[\"size_filter_large\"].astype(bool)) &\n",
    "    (xen_cropped[\"table\"].obs[\"size_filter_small\"].astype(bool)) &\n",
    "    (xen_cropped[\"table\"].obs[\"transcript_filter\"].astype(bool))\n",
    ")\n",
    "xen_cropped[\"table\"] = xen_cropped[\"table\"][mask].copy()\n",
    "xen_cropped[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10fc9dc2-c8b4-4940-ad22-356963e865ce",
   "metadata": {},
   "source": [
    "### 5. Data Normalisation, Dimensionality Reduction and Clustering\n",
    "\n",
    "As we have quite variable cell sizes and gene counts, we want to normalise these as before prior to clustering analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3551b625-1e53-47ff-b293-8028cf8f9aa6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].layers[\"counts\"] = xen_cropped[\"table\"].X.copy()\n",
    "sc.pp.normalize_total(\n",
    "    xen_cropped[\"table\"],\n",
    "    inplace=True\n",
    ")\n",
    "sc.pp.log1p(\n",
    "    xen_cropped[\"table\"]\n",
    ")\n",
    "xen_cropped[\"table\"] "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d34c1922-cd41-4e2c-a601-dfa76b6f4bb1",
   "metadata": {},
   "source": [
    "Principal Component Analysis (PCA) is a dimensionality reduction technique used to identify the primary axes of variation in high-dimensional data.\n",
    "In the context of spatial transcriptomics, PCA helps to reduce the complexity of the data while preserving the most important patterns of variation.\n",
    "\n",
    "**TIP:**\n",
    "If your target panel is very small, you can skip this step and carry out clustering analysis directly on gene expression.\n",
    "This can sometimes help with achieving better clustering results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fb0c444-2f48-4d95-8c11-15b595f41e63",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.pca(xen_cropped[\"table\"])\n",
    "xen_cropped[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdaa95a-22e1-4d72-b803-ac9a94184f93",
   "metadata": {
    "tags": []
   },
   "source": [
    "As before, we can visualise how much variation is captured by each PC.\n",
    "\n",
    "An elbow plot helps to determine the number of significant PCs to use for downstream analyses.\n",
    "\n",
    "The plot typically shows the amount of variance explained by each PC, and the “elbow” point indicates a natural cutoff."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c7d2fa8-6a82-44b7-a4d7-b6fc54fbccbf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.pca_variance_ratio(\n",
    "    xen_cropped[\"table\"],\n",
    "    n_pcs=50,\n",
    "    log=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4146350-da27-4404-8438-1411ed986213",
   "metadata": {},
   "source": [
    "Plotting the top genes contributing to a specific principal component helps in understanding the biological factors driving the variation captured by that component.\n",
    "This type of plot highlights the genes with the highest loadings, which are the most influential in the principal component analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bff328f-858b-43d0-8aac-e91c8ca4fadc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.pca_loadings(\n",
    "    xen_cropped[\"table\"],\n",
    "    components='1,2'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3074d2c-5ccb-4795-9c30-7fdf68a7f681",
   "metadata": {},
   "source": [
    "The `sc.pl.pca()` function in [scanpy](https://scanpy.readthedocs.io/en/stable/) is used to visualize the expression of a specific gene across cells in a given dimensionality reduction space (e.g., PCA).\n",
    "This helps to understand how the expression of a gene varies across the principal components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cb39f45-c0d9-4c13-84cf-6a6e44a83665",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.pca(\n",
    "    xen_cropped[\"table\"],\n",
    "    color=\"CEACAM5\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1989f7ab-34ec-45c4-a8a4-72205c484aa5",
   "metadata": {},
   "source": [
    "We can also examine how various PCs are distributed spatially.\n",
    "\n",
    "Here, we can see that high PC1 loadings enrich in crypt top cells and low PC1 loadings enrich in follicular structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31e7f4f2-98cf-4d29-b33c-0b561a710a7b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].obs[\"PC1\"] = xen_cropped[\"table\"].obsm[\"X_pca\"][:, 0]\n",
    "\n",
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"PC1\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"PC1\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc290652-961e-4d51-b0a4-b73af6a6644a",
   "metadata": {
    "tags": []
   },
   "source": [
    "We can plot the expression of high (or low) loading genes to visualise how this correlates with our dimensionality reduction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dde4b90-bca7-4bb8-892e-c2599c28faf4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"MS4A1\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"MS4A1 Expression\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a40824d4-2dc6-43bb-bd55-4491af663df6",
   "metadata": {},
   "source": [
    "Next, we will use the reduced dimensionality data for clustering and cluster visualisation.\n",
    "\n",
    "`sc.pp.neighbors()` needs to be run prior to `sc.tl.umap()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f5a9117-644e-41b2-ac0a-a0dcfc95c597",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.neighbors(\n",
    "    xen_cropped[\"table\"],\n",
    "    n_pcs=20\n",
    ")\n",
    "sc.tl.umap(\n",
    "    xen_cropped[\"table\"]\n",
    ")\n",
    "xen_cropped[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eac2fe8-8f79-4111-94c2-d82a5e66a243",
   "metadata": {},
   "source": [
    "Then we can also run `sc.tl.leiden()` to cluster cells using the Leiden algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1751049-f220-40f5-9ebe-ffdc48d57cf8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.leiden(\n",
    "    xen_cropped[\"table\"],\n",
    "    resolution=0.7,\n",
    "    key_added=\"cell_clusters\"\n",
    ")\n",
    "xen_cropped[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98282e87-fa86-4244-8ec4-69d817c3ea66",
   "metadata": {},
   "source": [
    "Next lets visualise the clusters - firstly, based on UMAP embedding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdd1675d-02b3-46ac-8280-f4b475b77c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color=\"cell_clusters\", \n",
    "    add_outline=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdf2b6ab-efb8-419a-bc04-0758a4e009d7",
   "metadata": {},
   "source": [
    "And now lets plot the clusters in tissue space.\n",
    "\n",
    "We can see that our clusters have quite nice correspondence to distinct spatial regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fd418b8-c1f1-42a6-89d3-05008efc7b09",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"cell_clusters\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Cell Clusters\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5aa479c7-2526-4193-a39b-be32423b646a",
   "metadata": {},
   "source": [
    "Sometimes the plots can get a bit busy if there are many clusters and the colours blend into each other.\n",
    "We can choose to visualise just a few groups at a time - for example, just clusters 1 and 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5be07cdc-130f-425c-8d3f-f5fb444903cb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"cell_clusters\",\n",
    "    groups=[\"1\", \"3\"]\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Selected Cell Clusters\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e26c16b1-fc19-4bf8-b9ad-f4b9614c0158",
   "metadata": {},
   "source": [
    "As before, now we can use [scanpy](https://scanpy.readthedocs.io/en/stable/) differential expression functions to identify marker genes for specific cell clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28f7e073-9703-45b8-b750-53fd9dcf1d9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(\n",
    "    xen_cropped[\"table\"],\n",
    "    'cell_clusters',\n",
    "    method='wilcoxon',\n",
    "    groups = ['5']\n",
    ")\n",
    "sc.pl.rank_genes_groups(\n",
    "    xen_cropped[\"table\"],\n",
    "    n_genes=25,\n",
    "    sharey=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce941bff-464e-431c-a7b6-97c74b2f2a36",
   "metadata": {},
   "source": [
    "We can visualise expression of cluster-specific markers using feature plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5306b9d7-b59b-4ab4-8215-e51781776c9f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color=\"CA4\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93652d24-c5bf-42cb-a7b0-92b7f8e2c673",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"CA4\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"CA4\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cac84fe1-891d-42b8-8336-4dff3832884a",
   "metadata": {},
   "source": [
    "`sc.pl.rank_genes_groups_dotplot()` provides a convenient and visually appealing way to display expression patterns of top marker genes across clusters using a dot plot.\n",
    "\n",
    "At this point, you can use any of the core [scanpy](https://scanpy.readthedocs.io/en/stable/) plotting functions with `xen_cropped[\"table\"]`. \n",
    "\n",
    "Explore various options and documentation here:\n",
    "\n",
    "https://scanpy.readthedocs.io/en/stable/tutorials/plotting/core.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "359ec496-0bb4-4915-9c0d-1bf2194d0f50",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(\n",
    "    xen_cropped[\"table\"],\n",
    "    groupby=\"cell_clusters\",\n",
    "    method=\"wilcoxon\"\n",
    ")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    xen_cropped[\"table\"],\n",
    "    n_genes=4\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "283727f7-fd1d-4a38-a754-3b7838e4ffdf",
   "metadata": {},
   "source": [
    "### 6. Cell Size Normalisation\n",
    "\n",
    "It has been demonstrated ([Atta et al, 2024](https://link.springer.com/article/10.1186/s13059-024-03303-w)), that if you have a very skewed panel design in in situ data, cell area/size normalisation can be more robust.\n",
    "A cell with low panel coverage would have lower counts for example, so normalising to “library size” would inflate those counts.\n",
    "Note that for large panels (e.g. Xenim 5K), it’s not such an issue.\n",
    "\n",
    "Let’s try normalising by area instead - there’s no in-built method is [scanpy](https://scanpy.readthedocs.io/en/stable/) for this, so we add the normalised counts and log transformed values (for visualisation) back to the object manually:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "536d7f01-340d-4c3e-bcb0-0aa8fa22caca",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "xen_cropped[\"table\"].layers['COUNT_NORM'] = xen_cropped[\"table\"].X\n",
    "raw_counts = xen_cropped[\"table\"].layers['counts']\n",
    "cell_area = xen_cropped[\"table\"].obs['cell_area'].values.astype(float) + 1e-6\n",
    "median_area = np.median(cell_area)\n",
    "inv = 1.0/cell_area\n",
    "D = sparse.diags(inv)\n",
    "norm_counts = D.dot(raw_counts)\n",
    "scaled_counts = norm_counts.multiply(median_area)\n",
    "log_norm_counts = np.log1p(scaled_counts)\n",
    "\n",
    "xen_cropped[\"table\"].layers['XENIUM_SIZE_NORM'] = log_norm_counts\n",
    "xen_cropped[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73734f07-27a0-4a33-b560-e41a846dc0ca",
   "metadata": {},
   "source": [
    "We can see that for some genes, there’s little difference - e.g. B-Cell marker *MS4A1*/*CD20*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9972afdd-f8ab-4481-b246-d77c31630b38",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plot_feature = \"MS4A1\"\n",
    "\n",
    "fig, ax = plt.subplots(1, 2, figsize = (12, 5))\n",
    "\n",
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color = plot_feature,\n",
    "    layer = \"XENIUM_SIZE_NORM\",\n",
    "    ax = ax[0],\n",
    "    show = False,\n",
    "    cmap=\"jet\"\n",
    ")\n",
    "\n",
    "ax[0].set_title('Area Size - XENIUM_SIZE_NORM')\n",
    "\n",
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color = plot_feature,\n",
    "    layer = \"COUNT_NORM\",\n",
    "    ax = ax[1],\n",
    "    show = False,\n",
    "    cmap=\"jet\"\n",
    ")\n",
    "\n",
    "ax[1].set_title(\"Transcript count - COUNT_NORM\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58c9eb04-9341-4d74-8231-bb9c03016474",
   "metadata": {},
   "source": [
    "But for others, standard normalisation and area size are very different:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e4e9ab-e340-44d3-b88d-387c69419105",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plot_feature = \"TUBB\"\n",
    "\n",
    "fig, ax = plt.subplots(1, 2, figsize = (12, 5))\n",
    "\n",
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color = plot_feature,\n",
    "    layer = \"XENIUM_SIZE_NORM\",\n",
    "    ax = ax[0],\n",
    "    show = False,\n",
    "    cmap=\"jet\"\n",
    ")\n",
    "\n",
    "ax[0].set_title('Area Size - XENIUM_SIZE_NORM')\n",
    "\n",
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color = plot_feature,\n",
    "    layer = \"COUNT_NORM\",\n",
    "    ax = ax[1],\n",
    "    show = False,\n",
    "    cmap=\"jet\"\n",
    ")\n",
    "\n",
    "ax[1].set_title(\"Transcript count - COUNT_NORM\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea37f1c5-196a-4142-a736-5ca9b9f74a14",
   "metadata": {},
   "source": [
    "You can see that the most difference correlates with large stromal cells.\n",
    "But, we can probably agree that due to nuclei-expansion approach to cell segmentation in this dataset, the area size of these cells is also likely to be overly inflated, and therefore area-based normalisation here is probably strongly under-estimating the true expression and creating a false gene expression pattern!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba6437fb-0072-419f-ac16-a6205cd9f38e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color = 'cell_area'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71342413-d836-4de3-914f-0623bf11c5b8",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 7. Cell type identification\n",
    "\n",
    "You can manually annotate your cell clusters, or you can classify them using a reference single-cell dataset.\n",
    "This process is simpler than for Visium data because our data is at the single-cell level, establishing a one-to-one relationship without the need for spot deconvolution.\n",
    "\n",
    "However, our transcriptome is more limited here, and some cell types may not be well represented. Additionally, our single-cell reference might be missing some cell types that are not well captured by droplet-based technologies but are present in our tissue data.\n",
    "\n",
    "In this example, we will use a single-cell reference dataset that we prepared earlier.\n",
    "\n",
    "We will start by reading in the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "552b6647-c6a9-4ac6-ba43-8ff676d54ca7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_ref = sc.read_h5ad(os.path.join(PRECOMPUTED_DIR, \"single_cell_colon_ref.h5ad\"))\n",
    "adata_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "960f5e09-5fce-4930-84e3-e85ba4933e67",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata_ref,\n",
    "    color = \"CellType\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa388fcd-f819-4680-9c23-4b0f54750faf",
   "metadata": {},
   "source": [
    "Before we proceed further, we subset the single cell dataset to the set of genes single cell and spatial have in common:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f804557-3b45-4610-b36e-eb4c1e849d4a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "common_genes = xen_cropped[\"table\"].var_names.intersection(adata_ref.var_names)\n",
    "adata_ref = adata_ref[:, common_genes].copy()\n",
    "len(common_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4505e36a-9a73-478c-b11e-e1aef5148e58",
   "metadata": {},
   "source": [
    "We want to evaluate how much structural information is lost in single-cell data when limiting ourselves to the targeted gene set.\n",
    "Accurate cluster prediction is challenging if the current gene set does not adequately identify them.\n",
    "To do this, we will quickly re-embedd the data using only the genes present in our spatial transcriptomics data and keep the original cluster annotations derived from unbiased data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edfa8b45-e602-46c3-a9e0-69a49a663520",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata_ref, inplace=True)\n",
    "sc.pp.log1p(adata_ref)\n",
    "\n",
    "sc.pp.pca(\n",
    "    adata_ref, \n",
    "    n_comps=50, \n",
    "    use_highly_variable=False, \n",
    "    svd_solver='arpack', \n",
    "    copy=False\n",
    ")\n",
    "\n",
    "\n",
    "sc.pp.neighbors(\n",
    "    adata_ref, \n",
    "    n_neighbors=20, \n",
    "    use_rep='X_pca',\n",
    "    n_pcs=20\n",
    ")\n",
    "\n",
    "sc.tl.umap(\n",
    "    adata_ref,\n",
    "    min_dist=0.3\n",
    ")\n",
    "\n",
    "adata_ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20969f47-0c42-4728-962d-7e9d9dfffb47",
   "metadata": {},
   "source": [
    "In this example, we can observe that the limited gene set does a reasonably good job at distinguishing major cell populations.\n",
    "However, it struggles to differentiate between similar cell types, such as myofibroblasts and fibroblasts, as effectively as before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a41cacb4-ad6e-414d-9a12-e51263f8445c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.embedding(\n",
    "    adata_ref,\n",
    "    basis=\"X_umap\",\n",
    "    color=\"CellType\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94a7bbc5-bf2c-4041-9e15-2c19be377e8d",
   "metadata": {},
   "source": [
    "If we visualise the specificity of the gene panel across our single cell reference clusters, we can see that the panel coverage is mainly concentrated across epithelial cells and T-Cells and other immune cells, with few specific markers expressed by stromal cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "815d8fe5-3847-43cb-9f9d-23467adae42c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_ref.layers[\"scaled\"] = sc.pp.scale(adata_ref, copy=True).X\n",
    "\n",
    "col_link = linkage(\n",
    "    pdist(\n",
    "        adata_ref.layers[\"scaled\"].T,\n",
    "        metric=\"euclidean\"\n",
    "    ),\n",
    "    method=\"average\"\n",
    ")\n",
    "col_order = leaves_list(col_link)\n",
    "\n",
    "mp = sc.pl.matrixplot(\n",
    "    adata_ref,\n",
    "    adata_ref.var_names[col_order],\n",
    "    \"CellType\",\n",
    "    dendrogram=True,\n",
    "    colorbar_title=\"mean z-score\",\n",
    "    layer=\"scaled\",\n",
    "    vmin=-1,\n",
    "    vmax=1,\n",
    "    cmap=\"RdBu_r\",\n",
    "    figsize=(11, 4),\n",
    "    show=False\n",
    ")\n",
    "mp[\"mainplot_ax\"].tick_params(\n",
    "    axis=\"x\",\n",
    "    labelsize=2,\n",
    "    bottom=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f1ac794-5c3b-421a-9931-4fd71d3fb75b",
   "metadata": {},
   "source": [
    "Next, we can use the standard `scanpy.tl.ingest()` integration and cross-classification workflow to transfer single-cell derived labels to our [SpatialData](https://spatialdata.scverse.org/en/stable/) object.\n",
    "\n",
    "Briefly, the function integrates embeddings and annotations of a new `adata` object with a reference dataset `adata_ref` through projecting on a PCA (or alternate model) that has been fitted on the reference data.\n",
    "The function uses a knn classifier for mapping labels and the UMAP package ([McInnes et al., 2018](https://arxiv.org/abs/1802.03426)) for mapping the embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5315d1e-14fd-4569-aa87-544990ac1ab4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.ingest(\n",
    "    xen_cropped[\"table\"],\n",
    "    adata_ref, obs='CellType',\n",
    "    embedding_method=['pca']\n",
    ")\n",
    "xen_cropped[\"table\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "462f6df1-6fb9-4b05-99f8-709b88008ab9",
   "metadata": {},
   "source": [
    "Unfortunately, the predicted labels and spatial clusters do not correspond clearly in all cases.\n",
    "This discrepancy is particularly evident in the middle regions of the UMAP, where many cells are predicted as epithelial cells - probably incorrectly!\n",
    "\n",
    "How to improve this?\n",
    "\n",
    "**Ensure Good Representation of Cell Type Markers in *in situ* Target Panel:**\n",
    "Most critically, before undertaking any experiments, you want to ensure that there is good representation of all cell types in your target panel.\n",
    "In this case, there is not much to be done as the data has already been generated.\n",
    "\n",
    "**Review and Refine Reference Data:**\n",
    "Ensure that the reference single-cell dataset is comprehensive and accurately annotated. If certain cell types are not well represented or annotated in the reference dataset, it can lead to misclassification.\n",
    "\n",
    "**Increase the Number of Dimensions:**\n",
    "Increasing the number of dimensions used in the UMAP and PCA steps might capture more variance in the data, leading to better label transfer.\n",
    "\n",
    "**Filter and Preprocess Data:**\n",
    "Filtering out low-quality cells or genes and performing additional preprocessing steps can enhance the accuracy of the transfer anchors and, consequently, the label predictions.\n",
    "\n",
    "**Manually Annotate or Correct Predictions:**\n",
    "In cases where automatic label transfer is insufficient, consider manually annotating or correcting the predictions for critical regions to ensure accuracy.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "582d555f-b1a8-4091-b187-8e3de3ebb902",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color = \"CellType\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "277e283e-fd11-4dee-b527-a4a7de8c13dc",
   "metadata": {},
   "source": [
    "As before, we can also visualise the predicted cell labels in tissue space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ad90f73-33da-4133-980a-0ca452083dd8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = xen_cropped.pl.render_labels(\n",
    "    element=\"cell_labels\",\n",
    "    table_name=\"table\",\n",
    "    fill_alpha=1,\n",
    "    color=\"CellType\"\n",
    ").pl.show(\n",
    "    coordinate_systems=\"global\",\n",
    "    title=\"Predicted Cell Type\",\n",
    "    return_ax=True\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64ed8ce0-93e0-4b53-8a5b-ea8a22bd2f0d",
   "metadata": {},
   "source": [
    "Unfortunately, `ingest()` does not provide any confidence score on its prediction.\n",
    "\n",
    "When other methods provide such scores, it can be helpful to visualise and diagnose issues such as:\n",
    "\n",
    "- lower confidence in certain spatial areas\n",
    "- lower confidence in cells embed \"between\" clusters (e.g. between B-cells and T-cells)\n",
    "\n",
    "This is often the case where cell segmentation is imperfect and partitions transcripts in such a way that it generates “artificial” doublets by pulling in transcripts from an adjacent cell.\n",
    "\n",
    "As an alternative to using our own single cell reference, we can use one of the reference models provided by [CellTypist](https://www.celltypist.org/).\n",
    "If you don't have a good reference of your own, this can be a quick and easy starting point.\n",
    "\n",
    "https://www.celltypist.org/\n",
    "\n",
    "First, let's download the available models and inspect them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1407f40e-d5d5-4af6-b2ed-a105fa0f37df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "models.download_models(\n",
    "    force_update = True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "443e9bf8-dc8b-4ed2-a38a-55d205686123",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "models.models_description()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9210f0e6-0145-4f44-9d53-64e621b2b482",
   "metadata": {},
   "source": [
    "To classify cells using [CellTypist](https://www.celltypist.org/) cell models, we simply need an [AnnData](https://anndata.readthedocs.io/en/stable/) object (normalised and log transformed) and a model name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5672cecb-8162-4bc9-a4f3-d6ea8477016c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata = xen_cropped[\"table\"].copy()\n",
    "adata.X = adata.layers[\"counts\"]\n",
    "sc.pp.normalize_total(\n",
    "    adata,\n",
    "    target_sum=10000,\n",
    "    inplace = True\n",
    ")\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32c23405-ce4f-45ae-8d33-55b8fde81bea",
   "metadata": {},
   "source": [
    "Here, we run predictions using human intestinal tract cells, which is probably the closest reference to human adult colon.\n",
    "You could also try the colorectal cancer reference here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf529993-c34f-4b38-a890-a686735bf254",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "predictions = celltypist.annotate(\n",
    "    adata,\n",
    "    model = 'Cells_Intestinal_Tract.pkl',\n",
    "    majority_voting = True\n",
    ")\n",
    "adata = predictions.to_adata()\n",
    "predictions.predicted_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d640a056-2121-4607-aa26-375f25a63d35",
   "metadata": {},
   "source": [
    "Now we can visualise predicted labels and also confidence scores.\n",
    "As with our previous method, we seem to be struggling to predict cell types in certain clusters. \n",
    "\n",
    "As previously covered, the stromal cells are likely poorly predicted due to limited panel coverage.\n",
    "What's interesting is that we often also see poor prediction confidence scores in-between clusters, for instance the division between T-cells and B-cells is not so clear, with low confidence cells forming at the cluster boundary.\n",
    "We know we should not be getting anything here that could be a biological \"intermediate\" cell, and therefore this hints at poor cell segmentation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "352437e4-845b-4e60-8f10-76d307f91794",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = 'conf_score'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d76928c-b742-49bd-8773-467aed55ba04",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color = 'majority_voting'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aae89c4-76d3-4fb2-b55e-981484d3dcd4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "celltypist.dotplot(\n",
    "    predictions,\n",
    "    use_as_reference = 'CellType',\n",
    "    use_as_prediction = 'majority_voting'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3de882e-dc74-4262-bbbe-21641529ee5d",
   "metadata": {
    "tags": []
   },
   "source": [
    "For example, if we visualise the lineage markers for T-Cells and B-Cells, we can see that they are often “co-expressed” in the same cells when biologically, they should not be.\n",
    "\n",
    "The `sc.pl.scatter()` function is used to create a scatter plot showing the relationship between the expression levels of two genes across all cells.\n",
    "This visualization helps to identify potential correlations or patterns between the two genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3655e406-ac46-411b-b2f7-d6dd6bafc44f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.scatter(\n",
    "    xen_cropped[\"table\"],\n",
    "    x=\"MS4A1\",\n",
    "    y=\"CD3D\",\n",
    "    color=\"CellType\",\n",
    "    size=20\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39a49c43-4c29-4743-918c-ce3653d27be6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    xen_cropped[\"table\"],\n",
    "    color = [\"MS4A1\", \"CD3D\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83429b76-e9f9-483f-b014-4164c4fd0015",
   "metadata": {},
   "source": [
    "### 8. Cell Re-segmentation \n",
    "\n",
    "To improve these artefacts, we can try alternative cell segmentation algorithms.\n",
    "What works best is very tissue dependant and there’s no easy one stop solution to this.\n",
    "Cell segmentation algorithms can be divided into a few groups.\n",
    "\n",
    "**Nuclei-based Segmentation** algorithms primarily focus on identifying cell nuclei, which are usually more distinct and easier to detect than the cell boundaries.\n",
    "Once the nuclei are identified, the cell boundaries are inferred by expanding around the nuclei.\n",
    "This approach works well in tissues where the nuclei are clearly visible and distinct and in early versions of many in situ platforms, were the only available methods due to only using DAPI stain.\n",
    "\n",
    "**Cell Boundary-Based Segmentation** algorithms (e.g. Cellpose) directly segments cells by identifying their boundaries.\n",
    "It is particularly effective for images with complex cell shapes and varying sizes, but this required good cell boundary staining - this is not available for our test dataset.\n",
    "Often cell boundary staining can be non-uniform across different tissues, adding further difficulties.\n",
    "Cellpose version 3 incorporates user-guided model training, which can be very useful for difficult to segment cell types - but this requires time investment to annotate training examples.\n",
    "\n",
    "**Transcript-Density Based Segmentation** algorithms, like Baysor segments cells based on the spatial distribution of transcripts\n",
    "It uses Bayesian inference to assign transcripts to cells, considering both the density and distribution of RNA molecules. This can be very useful for improving cell segmentation where cell boundary stain is not available or not working well.\n",
    "\n",
    "In this case, we will try re-segmenting our data with **Baysor**.\n",
    "Here’s the run we prepared earlier - see supplementary material on how to process the data yourself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5352a8d9-59b0-4957-b027-0b853961e4ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "baysor_dir = os.path.join(PRECOMPUTED_DIR, \"baysor\")\n",
    "\n",
    "seg = pd.read_csv(\n",
    "    os.path.join(baysor_dir, \"segmentation.csv\")\n",
    ")\n",
    "\n",
    "seg.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aa440c8-83c9-4ee1-accf-fd1a4ff2e189",
   "metadata": {},
   "source": [
    "There will be some transcripts that cannot be assigned to a cell - about 10% in this case.\n",
    "This information is stored under “is_noise” flag.\n",
    "This is fairly normal levels of noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f61dc70b-9446-4f29-8f4e-a3a16128973f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "seg['is_noise'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e880a01c-75fa-41e5-8ed6-d9339277cc51",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 5))\n",
    "plt.hist(\n",
    "    x=seg['assignment_confidence'],\n",
    "    bins=50\n",
    ")\n",
    "plt.xlabel('Seg Assignment Confidence')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "769547c0-bf5a-41ee-b2a5-a10a3b3d0794",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "(seg['assignment_confidence'] > 0.9).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae7e185-4b9d-4ba8-b338-121967b05440",
   "metadata": {},
   "source": [
    "plt.hist(x=seg['confidence'], bins=50)\n",
    "plt.xlabel('Seg Confidence')\n",
    "plt.show()And transcript confidence - the confidence that the molecule itself is real and not noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aed1023e-0b63-416b-8a97-87b8acc4f1e7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 5))\n",
    "plt.hist(\n",
    "    x=seg['confidence'],\n",
    "    bins=50\n",
    ")\n",
    "plt.xlabel('Seg Confidence')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3f357e8-8ef8-4197-82d8-5fcddaef5df0",
   "metadata": {},
   "source": [
    "We can filter out low confidence and low assignment confidence transcripts here from further analysis.\n",
    "How stringent you want to be depends on whether you want to keep as much data as possible and accept some inaccuracies, or end up with the cleanest possible dataset.\n",
    "\n",
    "Here, we will filter out transcripts that have not been assigned to cells, and below 0.9 confidence and assignment confidence.\n",
    "\n",
    "Then, we tabulate a cell by gene matrix from these data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eddcc5ed-3a45-4214-b6f4-6279352c0890",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "seg_filtered = seg[\n",
    "    (seg['confidence']> 0.9)&\n",
    "    (seg['assignment_confidence'] > 0.9) &\n",
    "    (~seg['is_noise'])\n",
    "]\n",
    "\n",
    "mat = pd.crosstab(\n",
    "    seg_filtered['gene'],\n",
    "    seg_filtered['cell']\n",
    ")\n",
    "mat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "059c1f08-a808-40bb-a0c7-d59f797391a3",
   "metadata": {},
   "source": [
    "[Baysor](https://kharchenkolab.github.io/Baysor/dev/) further provides diagnostic info about cells in `segmentation_cell_stats.csv` file, which we will also read in here.\n",
    "\n",
    "The following parameters can be used to filter low-quality cells:\n",
    "\n",
    "**area:**\n",
    "area of the convex hull around the cell molecules avg_confidence: average confidence of the cell molecules\n",
    "\n",
    "**density:**\n",
    "the number of molecules in a cell divided by the cell area\n",
    "\n",
    "**elongation:**\n",
    "ratio of the two eigenvalues of the cell covariance matrix\n",
    "\n",
    "**n_transcripts:**\n",
    "number of molecules per cell\n",
    "\n",
    "**avg_assignment_confidence:**\n",
    "average assignment confidence per cell.\n",
    "Cells with low avg_assignment_confidence have a much higher chance of being an artifact.\n",
    "\n",
    "**max_cluster_frac (only if n-clusters > 1):**\n",
    "fraction of the molecules coming from the most popular cluster.\n",
    "Cells with low max_cluster_frac are often doublets.\n",
    "\n",
    "**lifespan:**\n",
    "number of iterations the given component exists.\n",
    "The maximal lifespan is clipped proportionally to the total number of iterations.\n",
    "Components with a short lifespan likely correspond to noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc27df0a-852b-425c-af0e-1e369399872f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "stats = pd.read_csv(\n",
    "    os.path.join(baysor_dir,\"segmentation_cell_stats.csv\")\n",
    ")\n",
    "stats.set_index('cell', inplace=True)\n",
    "\n",
    "stats.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4854077b-479d-4a16-bcfc-f80b9aac1a90",
   "metadata": {},
   "source": [
    "Now, we can assemble an AnnData object as before:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32e9e2e1-9673-44f6-95d8-68ba22cc7e31",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_reseg = AnnData(mat.values.T)\n",
    "adata_reseg.var_names = mat.index\n",
    "adata_reseg.obs_names = mat.columns\n",
    "adata_reseg.obs = stats.loc[mat.columns]\n",
    "\n",
    "centroids = adata_reseg.obs[['x', 'y']].to_numpy()[:,[1,0]]\n",
    "centroids[:,1] = -centroids[:, 1]\n",
    "\n",
    "adata_reseg.obsm['spatial'] = centroids\n",
    "\n",
    "adata_reseg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b856dfe9-1352-4987-a9ca-641b5ee535b7",
   "metadata": {},
   "source": [
    "From here, we can use the seurat object to visualise various cell meta data - for example, average transcript assignment confidence per cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6875ccf-4b91-4dc4-aa62-a80f372b2895",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ax = sq.pl.spatial_scatter(\n",
    "    adata_reseg,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    size=2,\n",
    "    color=\"avg_assignment_confidence\",\n",
    "    figsize=(12, 6),\n",
    "    return_ax=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13152eaa-055a-4a6c-b776-1e6533764cf1",
   "metadata": {},
   "source": [
    "Let's filter out low count cells and re-cluster the data as before.\n",
    "In practice, you may also want to consider additional QC metrics here - e.g. removing overly small cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e399c2be-9e7a-43e9-bdb2-0d1df7bce4cc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(\n",
    "    adata_reseg,\n",
    "    inplace=True,\n",
    "    log1p=False\n",
    ")\n",
    "\n",
    "adata_reseg = adata_reseg[\n",
    "    adata_reseg.obs[\"total_counts\"] >= 15\n",
    "].copy()\n",
    "\n",
    "sc.pp.normalize_total(adata_reseg)\n",
    "sc.pp.log1p(adata_reseg)\n",
    "sc.pp.scale(adata_reseg)\n",
    "\n",
    "sc.pp.pca(\n",
    "    adata_reseg, \n",
    "    n_comps=25, \n",
    "    use_highly_variable=False, \n",
    "    svd_solver='arpack', \n",
    "    copy=False\n",
    ")\n",
    "\n",
    "sc.pp.neighbors(\n",
    "    adata_reseg, \n",
    "    n_neighbors=20, \n",
    "    use_rep='X_pca'\n",
    ")\n",
    "\n",
    "sc.tl.umap(\n",
    "    adata_reseg,\n",
    "    min_dist=0.3\n",
    ")\n",
    "\n",
    "sc.tl.leiden(\n",
    "    adata_reseg,\n",
    "    resolution = 0.5\n",
    ")\n",
    "\n",
    "adata_reseg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5108b15e-e938-4200-b95d-d06bdf476994",
   "metadata": {},
   "source": [
    "Visualising clusters, we can see that we already obtain a better separation in the UMAP embedding than before.\n",
    "Though of course, distances in the UMAP space can be very misleading and careful interpretation is required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b856c54e-cba6-4c81-a773-f7d4cc936224",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata_reseg,\n",
    "    color='leiden',\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ae40cb-279e-4132-9db6-de4b1a98cc56",
   "metadata": {},
   "source": [
    "Next we visualise the clusters in tissue space.\n",
    "As before, lets cross-classify our cells using the reference single cell dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c197066f-9f6f-482e-81a6-77ebe834a6e4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(12, 6))\n",
    "\n",
    "sc.pl.spatial(\n",
    "    adata_reseg,\n",
    "    color='leiden',\n",
    "    spot_size=10,\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ebc9857-c355-461b-9339-4d6b4f4ccaaf",
   "metadata": {},
   "source": [
    "As before, lets cross-classify our cells using the reference single cell dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4bf76a2-f23a-4602-8ee5-cb8bc4f6c517",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_reseg = adata_reseg[:, common_genes].copy()\n",
    "adata_ref = adata_ref[:, common_genes].copy()\n",
    "sc.tl.ingest(\n",
    "    adata_reseg,\n",
    "    adata_ref,\n",
    "    obs='CellType',\n",
    "    embedding_method=['pca']\n",
    ")\n",
    "\n",
    "adata_reseg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d619f19e-6886-4403-97ac-9f87dc3eab72",
   "metadata": {},
   "source": [
    "Visualising the predictions, we’ve separated T-Cells from B-Cells much better.\n",
    "The stromal clusters still predict poorly, but that is due to poor probe coverage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b362028-bc0f-4d21-853b-e12a16032a9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata_reseg,\n",
    "    color = 'CellType',\n",
    "    ax=ax,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "147165e2-d18c-4198-911a-39ada4568e35",
   "metadata": {},
   "source": [
    "We can check the distribution in tissue space:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be451cea-97ae-420c-8619-e1b7b3fcc0d2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "sc.pl.spatial(\n",
    "    adata_reseg,\n",
    "    color = \"CellType\",\n",
    "    spot_size=10,\n",
    "    ax=ax\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eee5de0-74ac-4eb5-8989-c7959b08012d",
   "metadata": {},
   "source": [
    "**Additional exercises**\n",
    "\n",
    "Congratulations on reaching the end of this notebook!\n",
    "\n",
    "Here are some additional prompts if you wish to practice more:\n",
    "\n",
    "- How does this compare with other segmentation algorithms?\n",
    "  For example, data resegmented with proseg has been made available here: \n",
    "\n",
    "```\n",
    "/project/shared/python/5_python_spatial_omics/data/precomputed/imaging/proseg\n",
    "```\n",
    "\n",
    "- What metrics would you use to quantify cell segmentation accuracy?\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
