{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "32b3d274-52fd-40d2-a9a6-6b3cd7ef83b1",
   "metadata": {},
   "source": [
    "# DAY 1: Visium Spatial Transcriptomics Data Analysis - Mouse Intestine\n",
    "\n",
    "## Overview\n",
    "\n",
    "This notebook will guide you through Day 1 content. We will cover:\n",
    "\n",
    "1. Loading data\n",
    "2. Quality control\n",
    "3. Normalisation, feature selection and dimensionality reduction\n",
    "4. Cluster analysis\n",
    "5. Spatial clustering \n",
    "\n",
    "## How to use this notebook\n",
    "\n",
    "This notebook is intended to be used as a reference for you own analysis.\n",
    "All code chunks have an explanation detailing the analysis steps and their purpose, as well as key parameters.\n",
    "Play around with these and see what they do, so that you are better equiped to adapt the workflow to your own data.\n",
    "\n",
    "## Dataset\n",
    "\n",
    "We will be using a Visium dataset from [Parigi et al, 2022](https://www.nature.com/articles/s41467-022-28497-0).\n",
    "This dataset was generated using V1 3' polyA Visium chemistry and consists of four mouse intestine samples taken from healthy mice and mice subjected to a DSS colitis model, where the intestine is damaged.\n",
    "Today, we will be using healthy mouse intestine section as an example.\n",
    "Overall, the dataset quality is good but there are some quality issues, which will hopefully show you what to look out for in your own data. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ac10cd4-1311-493f-b09c-d1139a8fdb37",
   "metadata": {},
   "source": [
    "## Setting Up Your Environment\n",
    "\n",
    "First we need to set up the environment and load the packages we will use for this workshop.\n",
    "\n",
    "**os:** Provides access to the operating system for handling files, directories and paths used in the notebook.\n",
    "\n",
    "**pandas:** Enables loading, cleaning and manipulating tabular data as dataframes.\n",
    "\n",
    "**matplotlib:** Provides core plotting functions for creating various visualisations and plots.\n",
    "\n",
    "**seaborn:** Builds on matplotlib for better plotting.\n",
    "\n",
    "**anndata:** AnnData is an established data structure for storing annotated single-cell datasets, which can also be used for spatial analysis.\n",
    "\n",
    "**scanpy:** Toolkit for preprocessing, analysing and visualising single-cell RNA-seq data that extends to spatial data.\n",
    "\n",
    "**squidpy:** Adds spatial analysis and visualization methods for spatial transcriptomics data, expanding on scanpy functionality.\n",
    "\n",
    "**spotsweeper:** A spatial QC package for detecting spatial outliers in spatial transcriptomics data.\n",
    "\n",
    "**banksy:** Performs spatially aware clustering of cells or spots using neighborhood information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc77b0e5-0fa5-4dbb-913f-58c0a17cb806",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Import os for working with system path\n",
    "import os\n",
    "# Import pandas for dataframe manipulation\n",
    "import pandas as pd\n",
    "# Import matplotlib and seaborn for plotting\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "# Import scanpy and anndata for single-cell RNA-seq\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "# Import squidpy for spatial transcriptomics\n",
    "import squidpy as sq\n",
    "# Import spotsweeper for quality control\n",
    "import spotsweeper.local_outliers as sw\n",
    "# import banksy for spatially aware clustering\n",
    "from banksy_utils.filter_utils import filter_hvg\n",
    "from banksy.main import median_dist_to_nearest_neighbour\n",
    "from banksy.initialize_banksy import initialize_banksy\n",
    "from banksy.embed_banksy import generate_banksy_matrix\n",
    "from banksy.main import concatenate_all\n",
    "from banksy_utils.umap_pca import pca_umap\n",
    "from banksy.cluster_methods import run_Leiden_partition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bb2d6d2-5957-4f9c-b13e-a39e1e90cbde",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Set Scanpy logging and plotting parameters\n",
    "sc.set_figure_params(facecolor=\"white\", figsize=(8, 8))\n",
    "sc.settings.verbosity = 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8b1debf-dd44-4a2c-bdb0-da1631e101dd",
   "metadata": {},
   "source": [
    "## 1. Loading Visium Spatial Transcriptomics Data\n",
    "\n",
    "We will start by loading Visium dataset.\n",
    "As for single cell data, community data analysis tools have a set of constantly evolving reader functions to ingest various spatial transcriptomics data formats.\n",
    "Unfortunately, how various platforms output and store their datasets is constantly evolving and therefore sometimes readers are not always working and up to date and you might require some workarounds if you have particularly new, or particularly old data!\n",
    "\n",
    "Here, we are using [squidpy](https://squidpy.readthedocs.io/en/stable/) package `read.visium()` function to ingest the data.\n",
    "[scanpy](https://scanpy.readthedocs.io/en/stable/) also has an equivalent `read_visium()` reader you can use, although it is now deprecated.\n",
    "\n",
    "`sq.read.visium()` call loads 10x Genomics Visium spatial transcriptomics data into an AnnData object, loading gene expression counts, H&E images and spatial metadata. \n",
    "\n",
    "#### Key arguments:\n",
    "\n",
    "**path:**\n",
    "Path to the 10x Visium output directory. Must contain the \"spatial\" folder and count matrix.\n",
    "\n",
    "**counts_file:**\n",
    "Name of the count matrix file to load - this will either be filtered_feature_bc_matrix.h5 or raw_feature_bc_matrix.h5. Filtered matrix will load only under tissue spots, while raw matrix contains data for all spots on the slide. It is always good to inspect this even if we don't use the outside tissue spots for the analysis.\n",
    "\n",
    "**library_id:**\n",
    "Identifier for the Visium section and needed for identifying samples when analysing more than one.\n",
    "\n",
    "**load_images:** (default: True)\n",
    "Whether to load histology images and scale factors into\n",
    "adata.uns[\"spatial\"][library_id].\n",
    "\n",
    "**source_image_path:**\n",
    "Optional path to the high-resolution tissue image if it is not in the standard location.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c2a641-1501-407e-ab20-5facdb952117",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "visium_dir = '/nvme/project/shared/python/5_python_spatial_omics/data/visium_v1_mouse_intestine/spaceranger/SRR14083626_HEALTHY_DAY0/outs'\n",
    "adata = sq.read.visium(visium_dir, counts_file='raw_feature_bc_matrix.h5', library_id=\"Day0\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81e6b98f-b2c8-41d8-b224-0d7e5a0b5aba",
   "metadata": {},
   "source": [
    "A warning above tells you that you have duplicate variable names (i.e. genes) in your data.\n",
    "This is because of mapping between Ensembl IDs and gene names is not one-to-one.\n",
    "For the purpose of the analysis, we should make the names unique. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebc5fba6-eec7-481b-8dd7-e2fb00ecbd7b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cdb5238-86df-44c3-9e34-a63a0d70dc9c",
   "metadata": {
    "tags": []
   },
   "source": [
    "Inspect the data structure.\n",
    "\n",
    "Below is a list of important \"fields\" of the AnnData structure, some of which are not immediately obvious from the summary view.\n",
    "\n",
    "**adata.X** – expression matrix\n",
    "\n",
    "**adata.obs** – spot metadata\n",
    "\n",
    "**adata.var** – gene metadata\n",
    "\n",
    "**adata.obs_names** - spot barcodes\n",
    "\n",
    "**adata.obsm[\"spatial\"]** – spot coordinates\n",
    "\n",
    "**adata.uns[\"spatial\"][\"Day0\"]** – images, scale factors, metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14d80b86-1d2e-4a9c-9697-fb2a968a91a5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfbd48f1-aded-43c9-86b5-0a0e58df85d7",
   "metadata": {},
   "source": [
    "Here we can see spot level meta data.\n",
    "At the moment, we only have the position of each spot (array_row & array_col), which is read in as part of the spot meta data.\n",
    "\n",
    "We can also see whether each spot is 'in_tissue' based on 10X Spaceranger's automated tissue detection based on the H&E image.\n",
    "\n",
    "Throughout the analysis, we will calculate additional metrics that will be stored here.\n",
    "It's a good place to stash any spot-level information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee55427f-8762-488e-a316-b5371bd7234f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "069bfb0d-649c-44c3-8d39-13cc88c56056",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c53d5a51-f189-4d05-8418-5864f203234a",
   "metadata": {
    "tags": []
   },
   "source": [
    "Here we can see feature (gene) meta data. We can also see that gene name is the primary index (that had to be made unique)\n",
    "\n",
    "Ensembl gene IDs are alaso included for the mm10 genome.\n",
    "\n",
    "Throughout the analysis, we will calculate additional gene level metrics that will get stored here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f980c445-2755-486c-bb9c-1bff597baed0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33ec572e-1915-4e7b-8f83-03bb132ce28a",
   "metadata": {
    "tags": []
   },
   "source": [
    "> **Exercise:**\n",
    "Try exploring other parts of the object.\n",
    "Where can you find coordinates and images?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c05747f5-edf0-4dd3-b7d8-18cfa6266c03",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02f6d2a-d3e5-4987-9f57-6b188fbe0fdc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obsm[\"spatial\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47874fcc-9e1a-40da-a38f-eaa5af8f18d0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.uns[\"spatial\"][\"Day0\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bb06274-c87a-433b-bdc4-73c2410f75f1",
   "metadata": {},
   "source": [
    "### Reading in Automated Tissue Detection Information\n",
    "\n",
    "10X Spaceranger software carries out automated tissue detection based on H&E images, which is typically reasonably good.\n",
    "The output is stored in the `tissue_positions.csv` text file.\n",
    "This is picked up automatically by squidpy, but could also be loaded manually using [pandas](https://pandas.pydata.org/). \n",
    "\n",
    "The provided code reads in this file and adds a metadata variable to the [AnnData](https://anndata.readthedocs.io/en/stable/) object.\n",
    "We also load the spatial coordinates.\n",
    "\n",
    "This can be useful if you have carried out tissue detection using a different software, for example, or if your data format is causing issues with default loaders. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ff5187f-fdcf-4f2a-b794-3f5eeb37d215",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load text file using pandas\n",
    "tissue_coords = pd.read_csv(os.path.join(visium_dir, \"spatial/tissue_positions.csv\"))\n",
    "tissue_coords = tissue_coords.set_index('barcode')\n",
    "tissue_coords = tissue_coords.loc[adata.obs_names]\n",
    "\n",
    "# Create new obs entry to indicate if each barcode is under tissue or not\n",
    "# by matching barcodes in obs_names with those in pandas dataframe\n",
    "adata.obs[\"under_tissue\"] = [\n",
    "    \"Under Tissue\"\n",
    "    if tissue_coords.loc[barcode, \"in_tissue\"]\n",
    "    else \"Outside Tissue\"\n",
    "    for barcode in adata.obs_names\n",
    "]\n",
    "\n",
    "# Lets also update the spatial coordinates in the object manually - \n",
    "# sometimes, depending on spaceranger versions,\n",
    "# your spaceranger spatial outputs may not be read in successfully by the default readers\n",
    "adata.obsm[\"spatial\"] = tissue_coords[[\"pxl_col_in_fullres\", \"pxl_row_in_fullres\"]].to_numpy()\n",
    "adata.obs[\"array_row\"] = tissue_coords[[\"array_row\"]]\n",
    "adata.obs[\"array_col\"] = tissue_coords[[\"array_col\"]]\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b773b68-3332-4d5d-813a-b3716feec117",
   "metadata": {},
   "source": [
    "Let's sanity check our data by quickly visualising the slide we loaded. \n",
    "\n",
    "The [squidpy](https://squidpy.readthedocs.io/en/stable/) `spatial_scatter()` function is used to create spatial plots of the spots on a tissue section.\n",
    "When an image such as H&E is available, all spatial data will be plotted over the image, which always provides a lot of useful information.\n",
    "Let's inspect the image first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "505fe7f0-aa79-4774-9803-e5538b6dbc7b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Visualise H&E staining for tissue\n",
    "# This can be dome using both Scanpy and Squidpy\n",
    "# sc.pl.spatial(adata, img_key=\"hires\")\n",
    "sq.pl.spatial_scatter(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "154730b5-af8a-4845-b393-5f61846478b8",
   "metadata": {},
   "source": [
    "If we provide a `colour=` arguement to the plotting function, we can plot additional variables stored in the object.\n",
    "Let's first check that the automated tissue detection that we read in matches the image. \n",
    "\n",
    "> **Exercise:**\n",
    "Do you think the tissue detection here is working well?\n",
    "Would you be happy with it for your own data?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d95ec472-5d83-4b2a-8836-95fa8a95786c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Visualization of tissue status in spatial coordinates\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"under_tissue\",\n",
    "    size=1.5\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfb098dc-abb6-4998-8b0f-7392c5abd8d7",
   "metadata": {},
   "source": [
    "## 2. Quality Control"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28c04f71-5606-4610-9cad-4bae1876f51e",
   "metadata": {},
   "source": [
    "The first thing we want to do after loading the data is quality control.\n",
    "We want to see how well the experiment has worked and what potential quality issues the data might have.\n",
    "As in single cell analysis, where we can have poor quality cells, in spatial experiments we can have lower quality spots, or lower quality regions (or the entire slide/section).\n",
    "Generally, we consider very similar metrics as in single cell analysis:\n",
    "\n",
    "- How many genes are detected in each spot?\n",
    "- How many UMIs are detected in each spot?\n",
    "- What's the complexity of each spot - do we have a lot of signal dominated by a handful of highly expressed genes?\n",
    "- What's the mitochondrial/ribosomal gene content of each spot?\n",
    "\n",
    "Unlike in single cell data however, the distribution of these metrics in spatial data tends to be a lot less uniform and there is no one uniform threshold that can be applied to every tissue uniformly.\n",
    "Tissue composition and cellular density in different regions can result in a lot of variability across these metrics, so you need to evaluate them carefully for the first time you work with any particular tissue. \n",
    "\n",
    "The code below calculates basic QC metrics using [scanpy](https://scanpy.readthedocs.io/en/stable/).\n",
    "You can add your own metrics - for example, you can look at the prevalence of gene signatures associated with cellular stress. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7638897a-a1e3-4d9d-bc98-a19fd961efd1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Identify mitochondrial and ribosomal genes\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\")\n",
    "adata.var[\"rb\"] = adata.var_names.str.startswith(\"Rp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73b648a7-9570-48fc-b03b-83753c0af5af",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Calculate QC metrics using scanpy\n",
    "sc.pp.calculate_qc_metrics(\n",
    "    adata,\n",
    "    qc_vars=[\"mt\", \"rb\"],\n",
    "    inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f55ade78-4213-4d7d-9ff1-a517c9f4fe4b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Inspect the updated object\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66e48d07-5940-4e24-b2af-839105931001",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Inspect the quality control data per spot\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5691e3c-e92e-4b80-bb82-8d3026e0f197",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Inspect the quality control data per gene\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93b1e0ca-1dc7-45d0-8d63-0cc96d9563f2",
   "metadata": {},
   "source": [
    "Next, we want to visualise the QC metrics.\n",
    "In this example, we visualise the distribution of the number of genes detected at each spot.\n",
    "This is a good first look overview metric to use to evaluate the quality of the data\n",
    "A high number of features may indicate areas with more complex or diverse cell types, while a low number might indicate poor-quality spots or regions with few active genes.\n",
    "\n",
    "In this example, we can see that the distributions for gene detection rate between under tissue and outside  tissue overlap - in an ideal dataset, this should not be happening!\n",
    "This can happen for several reasons: \n",
    "\n",
    "1. Tissue detection has not worked well and under tissue areas are misclassified as outside tissue areas. This can happen if you have tissue that does not stain well. \n",
    "2. Image is misaligned/rotated with respect to the data - check fiducial alignment in cellranger outputs, especially the corner fiducials\n",
    "3. There are issues with tissue permeabilisation - under-permeabilised tissues will have very low counts under tissue and higher counts outside tissue/at tissue edges, over-permeabilised tissue can result in transcript diffusion/leakage outside tissue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d532cee-4f20-4b2d-9b33-933528b3e42c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Use seaborn histogram function to plot number of genes\n",
    "sns.histplot(\n",
    "    adata.obs,\n",
    "    x=\"n_genes_by_counts\",\n",
    "    kde=True,\n",
    "    bins=60,\n",
    "    hue=\"under_tissue\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f998fe0-b4b7-4836-8b68-81b54b94c69d",
   "metadata": {},
   "source": [
    "Next, we visualise can visualise the spatial distribution.\n",
    "In this case, we can see that counts outside tissue area are observed mostly on one side of the tissue only, suggesting a technical issue with one side of the slide.\n",
    "This is a visium experiment using older V1 kits where the slide preparation is handled manually and we can end up with more technical variation (e.g. due to small pipetting errors) over the slide.\n",
    "This is less likely to occur in [CytAssist](https://www.10xgenomics.com/instruments/visium-cytassist) experiments, however evaluating the slides for issues is still critical. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "889edf47-cdb7-47de-9d70-e4de7ad88b76",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=[\"n_genes_by_counts\"],\n",
    "    cmap=\"jet\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02754c9e-c2c1-4b14-af9c-e35edf845308",
   "metadata": {},
   "source": [
    "Similar to the number of genes detected, we can also plot the number of molecules/UMIs detected per spot.\n",
    "The two metrics should generally correlate, but you may have tissue regions with homogenous cell types and therefore lower number of detected genes but you can still have high number of UMIs. \n",
    "\n",
    "Low molecule count and low gene count could mean that there are data quality issues in those regions.\n",
    "This could be due to tissue quality, or non-optimal tissue permeabilisation or other technical issues.\n",
    "\n",
    "But, it could also correspond to low cell density regions.\n",
    "It is important to understand the structural composition of your samples before you throw any data away.\n",
    "\n",
    "In spatial transcriptomics, there is generally much more variability in QC metrics between different regions than between different cell types in scRNA-Seq data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "812b77ad-a375-42ce-85cb-d47d14dd4fd0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Seaborn histogram\n",
    "sns.histplot(\n",
    "    adata.obs,\n",
    "    x=\"total_counts\",\n",
    "    hue=\"under_tissue\",\n",
    "    kde=True,\n",
    "    bins=60\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b64186d0-621c-4261-8276-f3ecb8b09300",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=[\"total_counts\"],\n",
    "    cmap=\"jet\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d4e8b50-5f82-4536-a912-d5d252de117b",
   "metadata": {},
   "source": [
    "As in single cell data, we can also evaluate the percentage mitochonrial and ribosomal genes per spot.\n",
    "Do not expect the distributions (and thresholds for filtering) to mimic single cell datasets perfectly (e.g. 10% mitochondrial threshold). \n",
    "\n",
    "We can see in the plots that there is a bias in mitochondrial and ribosomal gene expression in parts of the slide.\n",
    "\n",
    "The peanut-shaped region in the center of the slide with high ribosomal counts is an immune follicle.\n",
    "As T-cells generally express highly levels of ribosomal protein genes, it is likely due to this.\n",
    "\n",
    "We can also see higher percentage of mitochondrial genes in some parts of the slide, including non-tissue areas.\n",
    "Generally, we would expect epithelial cells to have a higher mitochondrial gene count, but in this case it does not quite correspond to histology across all areas.\n",
    "The left hand side and right hand side of the slide show shifts in these profiles, potentially indicating some technical artefacts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2953d996-0ec5-4a09-9d20-bc1ab17014d0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Seaborn violin plots for percentage mitochondrial and ribosomal RNA\n",
    "fig, axs = plt.subplots(1, 2, figsize=(15, 4))\n",
    "sns.violinplot(adata.obs[\"pct_counts_mt\"], ax=axs[0])\n",
    "sns.violinplot(adata.obs[\"pct_counts_rb\"], ax=axs[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "260f015f-a823-4e1b-976b-a18dab4473a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Seaborn histogram of percentage mitochondrial RNA inside and outside tissue\n",
    "sns.histplot(\n",
    "    adata.obs,\n",
    "    x=\"pct_counts_mt\",\n",
    "    hue=\"under_tissue\",\n",
    "    kde=True,\n",
    "    bins=60\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05316c05-914d-4184-99b7-7f5801673251",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Seaborn histogram of percentage ribosomal RNA inside and outside tissue\n",
    "sns.histplot(\n",
    "    adata.obs,\n",
    "    x=\"pct_counts_rb\",\n",
    "    hue=\"under_tissue\",\n",
    "    kde=True,\n",
    "    bins=60\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c52cb6d8-8be3-4809-9f91-f00146c9a587",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Spatial scatter of mitochondrial and ribosomal content\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=[\"pct_counts_rb\",\"pct_counts_mt\"],\n",
    "    cmap=\"jet\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65553ef8-716e-4813-8c2d-b38fad0df75b",
   "metadata": {},
   "source": [
    "### Spot complexity\n",
    "\n",
    "The code below visualises a complexity metric, indicating the percentage of total unique molecules occupied by the top 50 most highly expressed genes per spot in the spatial transcriptomics dataset.\n",
    "This can offer insights into the complexity and quality of each spot's transcriptome, particularly highlighting regions with very low diversity.\n",
    "\n",
    "A high percentage indicates that a small number of genes dominate the transcriptome in that spot, suggesting low complexity.\n",
    "Conversely, a lower percentage suggests a more diverse and complex transcriptome.\n",
    "\n",
    "As we would expect, the complexity of (most) spots outside the tissue covered area is very low and under tissue (mostly) high.\n",
    "So, although we often detect transcripts outside the tissue covered area, these tend to be dominated by few, highly abundant genes and are generally background.\n",
    "In this slide, this is not always the case though!\n",
    "\n",
    "We can use the values of QC metrics outside tissue as a guide to identify poor-quality spots under tissue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04bc3a60-8262-44c7-b1be-1f211c3d5529",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=[\"pct_counts_in_top_50_genes\"],\n",
    "    cmap=\"jet\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d65e736-c171-4015-870f-bc4d03df6e5a",
   "metadata": {},
   "source": [
    "Here we can see more clearly how the side of the tissue that has \"leaked\" has spot complexity comparable to that of spots under tissue. We should be very careful when looking at that area of the sample going forwards, as diffusion outside the tissue also means diffusion within the tissue in that area, which is harder to see and filter out. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27bee59a-d73a-4fac-b210-ba4033eb8c0f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Seaborn histogram of RNA complexity inside and outside tissue\n",
    "sns.histplot(\n",
    "    adata.obs,\n",
    "    x=\"pct_counts_in_top_50_genes\",\n",
    "    hue=\"under_tissue\",\n",
    "    kde=True,\n",
    "    bins=60\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba406a20-38c6-40a7-b4aa-af37b3c4a9e4",
   "metadata": {},
   "source": [
    "Inspecting the data allows us to make some decisions on the appropriate QC metrics and thresholds to use to clean up poor quality spots.\n",
    "However, unlike single cell data, a lot of the variation in spatial transcriptomics data is due to tissue architecture.\n",
    "We can expect a more acellular region to have fewer detected transcripts than a structure with dense or very transcriptionally active cells, which means applying uniform filtering thresholds across the whole slide could result in removal of good quality data. \n",
    "\n",
    "One alternative method we can use is tool called [SpotSweeper](https://pypi.org/project/spotsweeper/), which examines whether QC metrics for a spot are outliers without their local neighbourhood.\n",
    "This can be useful to flag up any technical issues - for example, sometimes we see systematic slide printing errors where probes in spots near slide edges will have very low counts, or spots in a very particular pattern.  \n",
    "\n",
    "We can examine individual QC metrics to flag local outliers. \n",
    "\n",
    "However, as the purpose of [SpotSweeper](https://pypi.org/project/spotsweeper/) is to flag local outliers, if there are quality issues with a large part/the entire slide, poor quality spots will not get flagged, so you cannot rely on it entirely. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ab1255-38ad-4e03-91a0-2d26f4209431",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sw.local_outliers(\n",
    "    adata,\n",
    "    metric=\"pct_counts_in_top_50_genes\",\n",
    "    direction=\"higher\",\n",
    "    n_neighbors=36,\n",
    "    sample_key=\"under_tissue\",\n",
    "    coord_key=\"spatial\",\n",
    "    log=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06c65034-b82f-4673-b923-b64b5adb56c6",
   "metadata": {},
   "source": [
    "Here we can see that we have a few low complexity spots within tissue surrounded by high complexityspots, so these get flagged as our local outliers. But the spots outside tissue which are surrounded by other low complexity spots are not flagged up. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44c8d279-1665-4db2-b9dc-5b7980f8c85e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"pct_counts_in_top_50_genes\",\n",
    "    cmap=\"jet\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6633e14f-566f-46e6-879d-c7ad58d0c9af",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs['pct_counts_in_top_50_genes_outliers_cat'] = adata.obs['pct_counts_in_top_50_genes_outliers'].astype('category')\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"pct_counts_in_top_50_genes_outliers_cat\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4514bcea-185f-4314-ab7b-f57d227b5325",
   "metadata": {},
   "source": [
    "> **Exercise:**\n",
    "> 1. How would you detect whether other QC metrics are local spatial outliers?\n",
    "> 2. Can you aggregate multiple metrics to flag spots that are outliers in more than one QC area? Can you include additional metrics in the below example?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6b3166c-6ae9-4d51-9306-192b6a1809f6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sw.local_outliers(\n",
    "    adata,\n",
    "    metric=\"pct_counts_mt\",\n",
    "    direction=\"higher\",\n",
    "    n_neighbors=36,\n",
    "    sample_key=\"under_tissue\",\n",
    "    coord_key=\"spatial\",\n",
    "    log=False\n",
    ")\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"pct_counts_mt\",\n",
    "    cmap=\"jet\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d805b6d-9b11-4573-9056-89693319e35b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs['pct_counts_mt_outliers_cat'] = adata.obs['pct_counts_mt_outliers'].astype('category')\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"pct_counts_mt_outliers_cat\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28d6681a-6f2e-4868-9852-8b774772f189",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs['all_metrics_outlier'] = (\n",
    "    adata.obs['pct_counts_mt_outliers'] | adata.obs['pct_counts_in_top_50_genes_outliers'] \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf17ab76-89c3-4461-beca-abb30545f883",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs['all_metrics_outlier_cat'] = adata.obs['all_metrics_outlier'].astype('category')\n",
    "sq.pl.spatial_scatter(adata, color=\"all_metrics_outlier_cat\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94251481-dd35-4079-818f-53845a1f4187",
   "metadata": {},
   "source": [
    "## 3. Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f53c80a-f4ae-4bb2-9756-d1fd112c151e",
   "metadata": {},
   "source": [
    "Once you have investigated and have a good idea of the quality and potential issues in your samples, we want to filter out areas of the slide that have technical artefacts, are poor quality or are outside of tissue. \n",
    "\n",
    "You can also exclude genes that are expressed in only a few spots, but that's generally not necessary.\n",
    "\n",
    "Below, we are removing spots that are:\n",
    "\n",
    "1. Outside tissue covered area\n",
    "2. Low complexity spots\n",
    "3. Spots with low gene counts\n",
    "4. Spots which are local outliers\n",
    "\n",
    "In this slide, you might also want to consider:\n",
    "\n",
    "1. Removing spots in tissue where there is a hair. Small debris is not uncommon in ST. \n",
    "2. Removing spots in the left hand side of the slide which appears to be affected a lot by permeabilisation issues. We will keep this for now to see how this affects downstream analysis!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67653caf-a5da-436f-9055-57f983d0691a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mask = (\n",
    "    (adata.obs[\"in_tissue\"] == True) &\n",
    "    (adata.obs[\"all_metrics_outlier\"] == False) &\n",
    "    (adata.obs[\"pct_counts_in_top_50_genes\"] < 30) &\n",
    "    (adata.obs[\"n_genes_by_counts\"] > 1000)\n",
    ")\n",
    "\n",
    "print(f\"Barcodes before filtering: {adata.n_obs}\")\n",
    "\n",
    "adata = adata[mask].copy()\n",
    "\n",
    "print(f\"Barcodes after cell count filter: {adata.n_obs}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b8a76c4-7ba7-4f0a-bf8e-a6de4b9550b3",
   "metadata": {},
   "source": [
    "A visual check on what we filtered out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad78e324-27bd-43b5-bebf-e582a4a50c8d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"under_tissue\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c1275c3-07f5-44e8-80b5-ad94ffa27e56",
   "metadata": {},
   "source": [
    "## 3. Normalisation, Feature Selection & Dimensionality Reduction\n",
    "\n",
    "Just as in the standard single-cell RNAseq data analysis workflow we now perform normalisation to adjust for spot library size. \n",
    "\n",
    "Lets pick a gene and see what it looks like before normalisation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13467a43-9874-427f-a46c-1c30d3995660",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"Myh11\",\n",
    "    title=\"Raw Counts Myh11\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a5eab12-3aca-432e-95e7-8c89eaa2ce1a",
   "metadata": {},
   "source": [
    "Normalise the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1daa8cb3-be59-4ff4-82e5-1f3179a6c1f2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Normalisation\n",
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "sc.pp.normalize_total(adata, inplace=True)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f77d6978-6ee1-4171-909b-2faa5ea9f12d",
   "metadata": {
    "tags": []
   },
   "source": [
    "Now we can see Myh11 values are a more uniform across the muscularis layers, not just concentrated in the center with high counts spots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c43817c-4a22-45da-ac27-d5ebcb88e28d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"Myh11\",\n",
    "    use_raw=False,\n",
    "    title=\"Normalised Myh11\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c7515d9-ce27-4fbe-a409-aa89c579201b",
   "metadata": {},
   "source": [
    "In unbiased ST experiments that measure the whole transcriptome, a lot of genes measured are not expressed, or are not differentially expressed or are generally uninformative.\n",
    "For downstream analysis, they add noise without contributing much information.\n",
    "Therefore, as in single cell data analysis, we want to select only a subset of genes which capture the most variability in the dataset for clustering analysis. \n",
    "\n",
    "NOTE - If you were working with targeted ST with smaller gene panels, this step is not required. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aabc587-6b48-4437-bc3e-b4315ebd6889",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Feature selection\n",
    "sc.pp.highly_variable_genes(\n",
    "    adata,\n",
    "    n_top_genes=2000\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b197efc-35e3-42c8-81d2-1080606e7697",
   "metadata": {},
   "source": [
    "We can now see that for each gene, we have calculated means and dispersions and added a flag 'highly_variable' on whether it is in the top highly variable genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9768ca67-c3cc-4f05-a87f-7731bb24f778",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03aad475-3834-4a43-ac4e-4b797d1f7e6a",
   "metadata": {},
   "source": [
    "The code below automatically identifies the most variable gene and plots its normalised expression:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81b5d658-ff8f-4491-bfdf-14e42f943251",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene = adata.var[\"dispersions_norm\"].sort_values(ascending=False).index[0]\n",
    "\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=gene,\n",
    "    use_raw=False,\n",
    "    title=f\"Most variable: {gene}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21eba49c-6191-42e8-a4ad-6062addd1b34",
   "metadata": {},
   "source": [
    "Reducing dimensionality is a crucial step in the analysis of high-dimensional spatial transcriptomics data.\n",
    "It simplifies the data structure while retaining its most informative features. \n",
    " \n",
    "Here, we'll run Principal Component Analysis (PCA) for dimensionality reduction.\n",
    "PCA reduces the dimensionality of the dataset by transforming it into a set of orthogonal axes, known as principal components (PCs), which capture the most variance in the data.\n",
    "By default, we'll store the top 50 PCs, which are typically more than enough to capture all the major axes of variation in spatial data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f947d798-92b4-460f-a1bd-cfcf8388c790",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# dimensionality reduction using PCA\n",
    "sc.pp.pca(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb21d0f3-04ea-495f-b6f0-f6bfe5410ef8",
   "metadata": {
    "tags": []
   },
   "source": [
    "After PCA, we need to decide how many PCs to use for clustering and further analyses.\n",
    "\n",
    "An *ElbowPlot* is used to select the number of PCs to include in downstream analyses.\n",
    "\n",
    "By plotting the percentage of variance explained by each PC, it helps identify the point at which additional PCs contribute minimal additional variance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e40a64e-f072-48d8-bf48-4a8e95a2d371",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Elbow plot\n",
    "sc.pl.pca_variance_ratio(\n",
    "    adata,\n",
    "    n_pcs=50,\n",
    "    log=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7567dade-74c2-4cc2-afc2-8b3ddf10399e",
   "metadata": {},
   "source": [
    "Biological relevance: Also consider biological knowledge and visual inspection of PCA plots to ensure that selected PCs capture relevant patterns.\n",
    "\n",
    "We can achieve this by visualising gene loadings of each PC to understand what kind of biological variation they capture.\n",
    "\n",
    " - PC 1: Might highlight major tissue regions or dominant cell types OR very dominant technical effects.\n",
    " - PC 10: Could reflect variation related to specific cellular processes or less prominent tissue structures.\n",
    " - PC 30: Might capture noise or subtle patterns, like rare cell states or technical variations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4156fad9-92a0-4e41-a09b-5962dfa523a7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.pca_loadings(adata, components=[1, 2, 3, 10, 20, 30])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6171b3e2-afa8-46fa-8e76-6437eb60879d",
   "metadata": {},
   "source": [
    "We have spatial data, so we can also see which regions in the slide load highly on which PC, which can help interpret whether PCs are capturing true biological variability or something technical. \n",
    "\n",
    "For example, we see that PC1 seems to be capturing the difference between mucosa and muscularis/submucosal layers, PC4 seems to correspond to lymphoid structures. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5847c56-e3a9-4045-b72d-49f59637d9bf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"PC1\"] = adata.obsm[\"X_pca\"][:, 0]\n",
    "sq.pl.spatial_scatter(adata, color=\"PC1\", cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff33f694-966d-4323-aa66-736bf1cabe81",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs[\"PC4\"] = adata.obsm[\"X_pca\"][:, 3]\n",
    "sq.pl.spatial_scatter(adata, color=\"PC4\", cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c987e91-b3cf-4ba2-b8ac-f33207056631",
   "metadata": {},
   "source": [
    "## 4. UMAP and Clustering\n",
    "\n",
    "Next, as in single cell analysis, we want to embedd the spots into two dimensional space to visualise which spots are similar to each other and which are different.\n",
    "\n",
    "Embedding spatial transcriptomics data using Uniform Manifold Approximation and Projection (UMAP) is a powerful way to visualise high-dimensional datasets in a two-dimensional space\n",
    "\n",
    "**Preserve Structure:**\n",
    "UMAP is designed to maintain both local and global data structures, which helps in visualising the relationships between data points accurately, although distances should be interpreted carefully. \n",
    "\n",
    "**Cluster Separation:**\n",
    "UMAP often provides better separation of clusters than other methods, making it useful for identifying distinct groups within data. \n",
    "\n",
    "**Scalability:**\n",
    "UMAP is computationally efficient, allowing it to handle large datasets commonly found in spatial transcriptomics.\n",
    "\n",
    "Here, the `n_pcs=` parameter is important - we want to select it based on the elbow plot.\n",
    "Too few PCs will miss important structures in the data, and too many will add too much noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bade3de-53d8-4801-ac47-86c765ba41a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# UMAP\n",
    "sc.pp.neighbors(\n",
    "    adata,\n",
    "    n_pcs=10\n",
    ")\n",
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "102c971d-bf8e-4ca9-bb2a-62ea319e5f5f",
   "metadata": {},
   "source": [
    "Next, we want to find clusters in the data.\n",
    "We will use leiden clustering, using the graph created in the previous step.\n",
    "The clustering algorithm partitions the spots into distinct groups.\n",
    "\n",
    "- The resolution parameter determines the granularity of the clustering.\n",
    "- *Resolution = 0.5* is a common starting point for many analyses, providing a balance between sensitivity and specificity.\n",
    "- However, the optimal resolution varies based on dataset complexity and biological context.\n",
    "- Higher resolution leads to more, smaller clusters. Useful for detailed analyses where subtle differences are biologically meaningful.\n",
    "- Lower resolution results in fewer, larger clusters. Suitable when you expect broader, more general differences in the data.\n",
    "- *n_iterations* How many iterations of the Leiden clustering algorithm to perform. Positive values above 2 define the total number of iterations to perform, -1 has the algorithm run until it reaches its optimal clustering. 2 is faster and the default for underlying packages.\n",
    "- *key_added* adata.obs variable name under which to add the cluster labels\n",
    "- *flavor* which packages implementation of the clustering algorythm to use. igraph is not the current default for compatability purposes, but it is the preffered method and the current default liedenalg will be deprecated shortly. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f84f348-c110-412b-b2cb-7278a89b1551",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Clustering\n",
    "sc.tl.leiden(\n",
    "    adata,\n",
    "    key_added=\"clusters\",\n",
    "    flavor=\"igraph\",\n",
    "    n_iterations=2,\n",
    "    resolution=0.5\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "570c1839-8892-4c4d-807c-f57dfebdbc78",
   "metadata": {},
   "source": [
    "We can now visualise detected clusters in UMAP space (based on transcriptional similarity) and tissue space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f1af8d6-2d52-45f3-b5db-7022fd6e5313",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=\"clusters\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87bd2c41-1505-4f75-b527-d211e55f8c71",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=\"clusters\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4688950-7f9b-4c0d-91a5-46c38b5fa456",
   "metadata": {},
   "source": [
    "##### The choice of cluster resolution significantly impacts the granularity of the resulting clusters.\n",
    "\n",
    "##### Let's explore the concept of clustering resolution and how to determine the optimal value for the analysis by trying high and low values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cedd488e-c707-4dc6-a6b2-f83ccb9529f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, key_added=\"clusters_res0.1\", flavor=\"igraph\", directed=False, n_iterations=2, resolution=0.1)\n",
    "sc.pl.umap(adata, color=\"clusters_res0.1\")\n",
    "sq.pl.spatial_scatter(adata, color=\"clusters_res0.1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6d2c156-93af-4085-9f88-af35d3e506bb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, key_added=\"clusters_res1\", flavor=\"igraph\", directed=False, n_iterations=2, resolution=1)\n",
    "sc.pl.umap(adata, color=\"clusters_res1\")\n",
    "sq.pl.spatial_scatter(adata, color=\"clusters_res1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0563747b-2a30-4881-af9c-1ddc83209399",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Marker Genes\n",
    "\n",
    "Next, we want to figure out what various clusters represent.\n",
    "To do so, we can find cluster marker genes. \n",
    "\n",
    "The `sc.tl.rank_genes_groups()` function is commonly used for this purpose, allowing you to perform differential expression analysis to identify genes that are significantly enriched in specific clusters compared to others.\n",
    "\n",
    "#### Wilcoxon Rank Sum Test (method = \"wilcoxon\")\n",
    "\n",
    "This non-parametric test is very commonly used, has been shown to perform well for detecting cluster marker genes and is suitable when no confounding variables need to be adjusted. \n",
    "\n",
    "NOTE: It is not suitable for detecting differentially expressed genes between biological groups with replicates, as it does not model individual sample effects in any way. For this, you want model-based or pseudobulk approaches which are more conservative.\n",
    "\n",
    "#### Student's t-test (method = \"t-test\")\n",
    "\n",
    "- **Parametric:** Assumes data is normally distributed.\n",
    "- Use Case: Suitable for large datasets where normality can be assumed.\n",
    " \n",
    "#### Student's t-test with over-estimated variance (method = \"t-test_overestim_var'\")\n",
    " \n",
    "Variant of t-test with overestimated variance that is more conservative\n",
    "\n",
    "#### Logistic Regression (method = \"logreg\")\n",
    "\n",
    "- Regression: Uses logistic regression to assess differential expression.\n",
    "- Use Case: Useful for handling confounding variables.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "304cc191-2ad6-4abd-b64e-d5b8fc010a7a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Identify cluster marker genes\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    \"clusters\",\n",
    "    method=\"wilcoxon\",\n",
    "    use_raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "535c1061-c5d1-43bb-a6c9-6d8205699cc2",
   "metadata": {},
   "source": [
    "The results table is stored as part of the unstructured data in the `AnnData` object, but that's not very useful for browsing the genes list!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cff4fd0-b8ff-4883-938e-7e877f227456",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.uns[\"rank_genes_groups\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "320ee7f3-5206-45b4-983f-972699710194",
   "metadata": {},
   "source": [
    "We can visualise it using a scanpy helper function sc.get.rank_genes_groups_df. For example, fetch genes for cluster 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb7366de-efaa-4a71-ac7f-1a4f2e8c06a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.get.rank_genes_groups_df(\n",
    "    adata,\n",
    "    group=\"1\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "480441b2-0b31-46cc-bbb6-7a95c58ca4c0",
   "metadata": {},
   "source": [
    "We can also use standard [scanpy](https://scanpy.readthedocs.io/en/stable/) plotting functions to plot heatmaps to visualise how gene groups differ between clusters.\n",
    "For example, here we are visualisig top 5 genes per cluster. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed16e81d-54f8-47b0-bac6-2c8a82820fb4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata,\n",
    "    n_genes=5,\n",
    "    groupby=\"clusters\",\n",
    "    standard_scale=\"var\",\n",
    "    show_gene_labels=True,\n",
    "    use_raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b00d3d6b-8dcd-49a2-92aa-155869659d6d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Visualization of cluster markers in spatial coordinates\n",
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=[\"clusters\", \"Muc2\"],\n",
    "    use_raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "836c5b47-2dd4-412c-9f40-48de40877032",
   "metadata": {},
   "source": [
    "## 5. Spatial Clustering\n",
    "\n",
    "So far in our analysis, we have not used spatial information (outside QC). \n",
    "\n",
    "As you can appreciate, single cell type workflows can get you very far in spatial transcriptomics analysis, and transcriptome-only clustering appears to correspond to histology and tissue structure reasonably well. \n",
    "\n",
    "However, there are several ways to incorporate spatial information that could yield better results here. \n",
    "\n",
    "Typically, spatial clustering approaches fall into broad categories:\n",
    "\n",
    "1. Expression-only clustering -\n",
    "   no spatial information considered, this is the workflow we did above.\n",
    "   Fast and no additional bells and whistles.\n",
    "2. Expression-only clustering with spatially informed feature selection. \n",
    "3. Graph-based spatial clustering -\n",
    "   build a graph using both expression similarity and spatial adjacency,\n",
    "   then cluster the graph. \n",
    "4. Spatially smoothed / regularized expression clustering -\n",
    "   cluster on expression, but penalise spatial discontinuity.\n",
    "5. Joint expression–spatial embeddings -\n",
    "   Learn a latent space combining expression + coordinates,\n",
    "   typically with deep models or graph embeddings that ingest both.\n",
    "6. Image-informed spatial clustering -\n",
    "   incorporating additional features from H&E image\n",
    "\n",
    "Next, we will explore some of these but this is a very active field of development and many alternatives are available."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b4b2ecf-6a3a-4133-8339-b6559a0f2133",
   "metadata": {},
   "source": [
    "### Spatially variable genes\n",
    "\n",
    "We will first try a simple extension to our previous approach: expression-only clustering with spatially informed feature selection.\n",
    "\n",
    "Previously, we calculated variable features as input without considering spatial information. \n",
    "Here, we will modify the workflow and use spatially variable genes instead. \n",
    "\n",
    "Spatial variation can be caused by differences in cell-type composition, overall functional dependencies, or cell-cell communication events, and help to understand the underlying tissue biology.\n",
    "Methods designed to identify spatially variable genes (SVGs) are designed to quantify whether a gene shows a significant spatial pattern by typically decomposing spatial and non-spatial variation in the dataset.\n",
    "Several methods have been proposed for this task with varying complexity and different assumptions.\n",
    "Currently there is no consensus on which method works best and how to define spatial variability in general. \n",
    "\n",
    "- SpatialDE (Svensson et al., 2018), SpatialDE2 (Kats et al., 2021) and SPARK (Zhu et al., 2021; Sun et al., 2020) use spatial correlation testing. \n",
    "- Sepal (Andersson and Lundeberg, 2021) leverages a Gaussian diffusion on spatial expression.\n",
    "- scGCO (Zhang et al., 2022) utilizes a graph cut method.\n",
    "- SpaGCN (Hu et al., 2021) identifies SVGs based on spatial domains identified through a graph convolutional neural network.\n",
    "\n",
    "In this tutorial, we will use one of the most commonly used approaches, Moran's I. \n",
    "\n",
    "### Moran’s I \n",
    "\n",
    "- Moran's I is a measure of spatial autocorrelation\n",
    "- It quantifies the degree to which similar values occur near each other in a spatial dataset.\n",
    "- It is widely used in spatial analysis to determine if there is a pattern in the spatial distribution of a particular variable, such as gene expression levels in spatial transcriptomics data.\n",
    "- Moran's I helps identify spatially variable features that may be linked to biological structures or processes.\n",
    "\n",
    "We can calculate Moran's I in [squidpy](https://squidpy.readthedocs.io/en/stable/) by first identifying spatial neighbours and then calling the `spatial_autocorr()` function.\n",
    "The outputs are stored as part of the unstructred data in [AnnData](https://anndata.readthedocs.io/en/stable/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a160eac1-7983-4b9f-9972-88e92d575b9d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.gr.spatial_neighbors(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "827ab6db-2316-49c4-b10a-56d14b1b49f9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.gr.spatial_autocorr(\n",
    "    adata,\n",
    "    mode=\"moran\",\n",
    "    genes=adata.var_names\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47d1c1aa-a74e-47a7-808f-aa460612afad",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.uns[\"moranI\"].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "799472dc-43a7-4f91-a41e-9253e1bdb18a",
   "metadata": {},
   "source": [
    "We can visualise the spatial expression of the top most spatially variable features to check that they are indeed spatially spread out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44f136c4-fccc-4afe-8bfc-420152766fbc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=[\"mt-Co3\", \"Muc2\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b41c2798-f75d-435e-9aac-7a749431f5e3",
   "metadata": {},
   "source": [
    "> **Exercise:**\n",
    "What's the difference between variable genes and spatially variable genes? In this case we can see that a lot of genes with high dispersion we identified earlier have low spatial variability and vice versa. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66cea7b9-fba1-4aee-bc15-f2d40eb21641",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame({\n",
    "    \"moranI\": adata.uns[\"moranI\"][\"I\"],\n",
    "}).join(\n",
    "    adata.var[[\"dispersions_norm\"]]\n",
    ")\n",
    "\n",
    "plt.scatter(df[\"moranI\"], df[\"dispersions_norm\"], s=5)\n",
    "plt.xlabel(\"Moran's I\")\n",
    "plt.ylabel(\"Dispersion\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "947c1a95-3e39-4106-8395-d2af58b7d518",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# select top 2000 genes by Moran’s I\n",
    "gene_list = (\n",
    "    adata.uns[\"moranI\"][\"I\"]\n",
    "    .sort_values(ascending=False)\n",
    "    .head(2000)\n",
    "    .index\n",
    ")\n",
    "\n",
    "# flag genes for PCA\n",
    "adata.var[\"use_for_pca\"] = adata.var_names.isin(gene_list)\n",
    "\n",
    "# run PCA on those genes only\n",
    "sc.tl.pca(adata, mask_var=\"use_for_pca\")\n",
    "# run neighbour detection and clustering\n",
    "\n",
    "sc.pp.neighbors(adata, n_pcs=10)\n",
    "sc.tl.leiden(adata, key_added=\"clusters_morans\", flavor=\"igraph\", directed=False, n_iterations=2, resolution=0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "573df752-c3b2-4374-b1cb-0fcd8850195c",
   "metadata": {},
   "source": [
    "Lets visualise - the difference is not very big between two clustering solutions, but spatially variable genes tend to give us a more spatially contiguous solution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfa926e2-89be-4b83-b031-b67d2d8d3b35",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=[\"clusters_morans\", \"clusters\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "550f26d5-16f7-4652-b38c-071ef39b03e0",
   "metadata": {},
   "source": [
    "### Banksy\n",
    "\n",
    "The second spatial clustering method we will try is Banksy algorithm, which performs consistently well across multiple benchmarks. \n",
    "\n",
    "https://www.nature.com/articles/s41588-024-01664-3\n",
    "\n",
    "Banksy is a spatial clustering algorithm that integrates both gene expression and spatial proximity. \n",
    "\n",
    "It works by augmenting each spot’s gene expression with information from its spatial neighbours, then applying standard dimensionality reduction and clustering.\n",
    "\n",
    "Conceptually, it does the following:\n",
    "\n",
    "1. **Start from expression features:** each spot has a vector of gene expression values\n",
    "\n",
    "2. **Build a spatial neighbourhood graph:** for every spot, BANKSY identifies nearby spots using spatial coordinates and distance-based neighbours.\n",
    "\n",
    "3. **Aggregate neighbour expression:** for each spot, it computes a weighted average of its neighbours’ expression, where weights decay with distance.\n",
    "\n",
    "4. **Augment the feature space:** the original expression and the neighbour-aggregated expression are combined into a new feature representation, controlled by a mixing parameter lambda that sets how strongly spatial context influences each spot.\n",
    "\n",
    "5. **Run standard embedding and clustering:** PCA is applied to the augmented features, followed by graph-based clustering.\n",
    "\n",
    "Because each spot’s features already encode local spatial context, clusters tend to form contiguous tissue domains without explicitly enforcing spatial smoothness in the clustering objective.\n",
    "\n",
    "In short, BANKSY turns expression + neighbourhood expression into a single feature matrix, so spatial structure is baked into the data before clustering.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8c698f7-20ab-4fc1-866c-63b4d5b797b8",
   "metadata": {},
   "source": [
    "Set Banksy parameters.\n",
    "\n",
    "Two critical parameters to consider are `lambda` and `k_geom`.\n",
    "These control how far around each spot we are looking and how much weight is given to spot vs neighbourhood features for clustering.\n",
    "\n",
    "Conceptually, this means the difference between some spatial \"smoothing\" (low `k_geom`, low `lambda`) and detection of broader spatial structures (high `k_geom`, high `lambda`). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c493c099-d36f-423d-8de1-339361e63f68",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "k_geom = 15 \n",
    "lambda_list = [0.2] "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09d20aca-b032-4521-a756-59c16a0294a8",
   "metadata": {},
   "source": [
    "Select top highly variable genes from your previous [AnnData](https://anndata.readthedocs.io/en/stable/) object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b9c668-08b1-4b2b-b7d2-db8730d83fd7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_banksy = adata[:, adata.var[\"highly_variable\"]].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28cab623-6e4e-485f-8244-a0427b40eed4",
   "metadata": {},
   "source": [
    "Find the median distance to closest neighbours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1541f6d5-a20e-4203-bf53-7f9c3b7174b9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from banksy.main import median_dist_to_nearest_neighbour\n",
    "nbrs = median_dist_to_nearest_neighbour(\n",
    "    adata_banksy,\n",
    "    key = \"spatial\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36a404b0-a28e-4c0b-8c8e-b651bd69c2f8",
   "metadata": {},
   "source": [
    "Initialise banksy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36ea6818-a9d3-41fc-939d-0b93254333f6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from banksy.initialize_banksy import initialize_banksy\n",
    "banksy_dict = initialize_banksy(\n",
    "    adata_banksy,\n",
    "    ('xcoord', 'ycoord', 'spatial'),\n",
    "    k_geom,\n",
    "    nbr_weight_decay=\"scaled_gaussian\",\n",
    "    max_m=1,\n",
    "    plt_edge_hist=False,\n",
    "    plt_nbr_weights=False,\n",
    "    plt_agf_angles=False,\n",
    "    plt_theta=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd421ab1-6d2a-41e4-b657-a4bdbf59a382",
   "metadata": {},
   "source": [
    "Generate `banksy matrix`, which calculates neighbourhood weighted gene expression features for selected parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51e4005e-44a2-4990-b761-c712b883bcc3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from banksy.embed_banksy import generate_banksy_matrix\n",
    "banksy_dict, banksy_matrix = generate_banksy_matrix(\n",
    "    adata_banksy,\n",
    "    banksy_dict,\n",
    "    lambda_list,\n",
    "    1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cfe090e-8b80-4481-8718-d33ff4b0ec98",
   "metadata": {
    "tags": []
   },
   "source": [
    "We can now use this matrix as input for PCA, UMAP and clustering. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f45a5313-e93d-4dc4-b5c0-cdf45eecb220",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from banksy_utils.umap_pca import pca_umap\n",
    "pca_umap(\n",
    "    banksy_dict,\n",
    "    pca_dims=[20],\n",
    "    add_umap=True,\n",
    "    plt_remaining_var=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34f28a60-b505-4c51-9658-9acb26a4cbe3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from banksy.cluster_methods import run_Leiden_partition\n",
    "results_df, max_num_labels = run_Leiden_partition(\n",
    "    banksy_dict,\n",
    "    [0.7],\n",
    "    num_nn=50,\n",
    "    num_iterations=-1,\n",
    "    partition_seed=1234,\n",
    "    match_labels=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd488706-ece1-4028-94c3-67c749789719",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata.obs['clusters_banksy'] = results_df.relabeled[0].dense\n",
    "adata.obs['clusters_banksy'] = adata.obs['clusters_banksy'].astype(\"category\")\n",
    "sq.pl.spatial_scatter(adata, color='clusters_banksy')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfe71d5e-4f00-4931-91cc-34aa8dd236f7",
   "metadata": {},
   "source": [
    "> **Exercise:**\n",
    "How does the Banksy clustering solution compare to previous ones?\n",
    "Try varying parameter settings to understand how to focus the data partitioning more on space or more on individual spot transcriptome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59bb1b65-73d0-492f-ac52-8f3178a2280e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    color=['clusters_banksy', \"clusters_morans\", \"clusters\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38c139e7-69ea-4863-9118-1ecc4aca47b1",
   "metadata": {},
   "source": [
    "### Saving the processed object for further analysis\n",
    "\n",
    "Save your [AnnData](https://anndata.readthedocs.io/en/stable/) object as a [h5ad](https://anndata.readthedocs.io/en/latest/fileformat-prose.html) file.\n",
    "You can load this back at any time to continue the analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "330848f1-51e7-4caf-a2cd-f4fb5086ca9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set location to store analysis output\n",
    "OUTPUT_FOLDERNAME = \"/nvme/project/USERNAME/PATH_TO_DAY1/\"\n",
    "adata.write(os.path.join(OUTPUT_FOLDERNAME, 'day1.h5ad'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
